<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Sparse regression with paired covariates</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Sparse regression with paired covariates</h1>



<p>Code for reproducing the results shown in the manuscript.</p>
<div id="initialisation" class="section level1">
<h1>Initialisation</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### Loading functions. <span class="al">###</span>

inst &lt;-<span class="st"> </span><span class="kw">rownames</span>(utils<span class="op">::</span><span class="kw">installed.packages</span>())

cran &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;devtools&quot;</span>,<span class="st">&quot;R.utils&quot;</span>,<span class="st">&quot;Matrix&quot;</span>,<span class="st">&quot;glmnet&quot;</span>,<span class="st">&quot;pROC&quot;</span>,<span class="st">&quot;BiocManager&quot;</span>,<span class="st">&quot;ashr&quot;</span>)
<span class="co"># &quot;googledrive&quot;, &quot;httpuv&quot;</span>
<span class="cf">if</span>(<span class="op">!</span><span class="kw">all</span>(cran <span class="op">%in%</span><span class="st"> </span>inst)){
    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(cran)){
        <span class="cf">if</span>(<span class="op">!</span>cran[i] <span class="op">%in%</span><span class="st"> </span>inst){
            <span class="kw">install.packages</span>(cran[i])
        }
    }
}

bioc &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;edgeR&quot;</span>,<span class="st">&quot;TCGAbiolinks&quot;</span>)
<span class="cf">if</span>(<span class="op">!</span><span class="kw">all</span>(bioc <span class="op">%in%</span><span class="st"> </span>inst)){
    <span class="co">#source(&quot;http://bioconductor.org/biocLite.R&quot;)</span>
    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(bioc)){
        <span class="cf">if</span>(<span class="op">!</span>bioc[i] <span class="op">%in%</span><span class="st"> </span>inst){
            <span class="co">#biocLite(bioc[i])</span>
            BiocManager<span class="op">::</span><span class="kw">install</span>(bioc[i])
        }
    }
}

<span class="co">#if(!&quot;ashr&quot; %in% inst){</span>
<span class="co">#  devtools::install_github(&quot;stephens999/ashr&quot;)</span>
<span class="co">#}</span>

user &lt;-<span class="st"> </span><span class="kw">Sys.getenv</span>(<span class="st">&quot;USERNAME&quot;</span>)
path &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="st">&quot;C:&quot;</span>,<span class="st">&quot;Users&quot;</span>,user,<span class="st">&quot;Desktop&quot;</span>,<span class="st">&quot;palasso&quot;</span>)
<span class="cf">if</span>(user<span class="op">==</span><span class="st">&quot;arra&quot;</span>){path &lt;-<span class="st"> &quot;C:/Users/arra/Desktop/MATHS/palasso_desktop&quot;</span>}
<span class="cf">if</span>(user<span class="op">==</span><span class="st">&quot;&quot;</span>){path &lt;-<span class="st"> &quot;/virdir/Scratch/arauschenberger/palasso&quot;</span>}
<span class="kw">setwd</span>(path)
folders &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;data&quot;</span>,<span class="st">&quot;results&quot;</span>)
<span class="kw">invisible</span>(<span class="kw">sapply</span>(folders,<span class="cf">function</span>(x) <span class="cf">if</span>(<span class="op">!</span><span class="kw">dir.exists</span>(x)){<span class="kw">dir.create</span>(x)}))

<span class="cf">if</span>(user<span class="op">!=</span><span class="st">&quot;arra&quot;</span>){
    devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;kkdey/CorShrink&quot;</span>) <span class="co"># ref=&quot;a9f6ba0&quot;</span>
    devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;rauschenberger/palasso&quot;</span>) <span class="co"># ref=&quot;4a995a2&quot;</span>
}

<span class="cf">if</span>(<span class="ot">FALSE</span>){
    
    <span class="co"># The functions &lt;&lt;save&gt;&gt;, &lt;&lt;file.exists&gt;&gt; and &lt;&lt;file.remove&gt;&gt; access the hard disk, but also try to access googledrive.</span>
    
    save &lt;-<span class="st"> </span><span class="cf">function</span>(object,file){
        base<span class="op">::</span><span class="kw">save</span>(object,<span class="dt">file=</span>file)
        <span class="kw">tryCatch</span>(<span class="dt">expr=</span>googledrive<span class="op">::</span><span class="kw">drive_upload</span>(<span class="dt">media=</span>file,<span class="dt">path=</span>file),
                 <span class="dt">error=</span><span class="cf">function</span>(x) <span class="ot">NULL</span>)
        <span class="co">#Sys.sleep(0.5)</span>
    }

    file.exists &lt;-<span class="st"> </span><span class="cf">function</span>(file){
        offline &lt;-<span class="st"> </span>base<span class="op">::</span><span class="kw">file.exists</span>(file)
        online &lt;-<span class="st"> </span><span class="ot">FALSE</span>
        <span class="cf">if</span>(<span class="op">!</span>offline){
            d &lt;-<span class="st"> </span>googledrive<span class="op">::</span><span class="kw">as_dribble</span>(<span class="dt">x=</span>file)
            online &lt;-<span class="st"> </span><span class="kw">tryCatch</span>(<span class="dt">expr=</span>googledrive<span class="op">::</span><span class="kw">some_files</span>(d),
                               <span class="dt">error=</span><span class="cf">function</span>(x) <span class="ot">FALSE</span>)
            <span class="co">#Sys.sleep(0.5)</span>
        }
        <span class="kw">return</span>(offline<span class="op">|</span>online)
    }

    file.remove &lt;-<span class="st"> </span><span class="cf">function</span>(file){
        base<span class="op">::</span><span class="kw">file.remove</span>(file)
        <span class="kw">tryCatch</span>(<span class="dt">expr=</span>googledrive<span class="op">::</span><span class="kw">drive_trash</span>(<span class="dt">file=</span>file),
             <span class="dt">error=</span><span class="cf">function</span>(x) <span class="ot">NULL</span>)
        <span class="co">#Sys.sleep(0.5)</span>
    }
    
    
<span class="co"># The function &lt;&lt;update&gt;&gt; moves files from the research cloud to the remote drive. Given both paths, it first verifies whether the folders SIM, GDC and CCA are available, and then copies all missing files from the research cloud to the remote drive.</span>

<span class="co"># from: path to the origin</span>
<span class="co"># to: path to the destination</span>

update &lt;-<span class="st"> </span><span class="cf">function</span>(from,to){
    dir &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;SIM&quot;</span>,<span class="st">&quot;GDC&quot;</span>,<span class="st">&quot;CCA&quot;</span>)
    <span class="cf">if</span>(<span class="kw">any</span>(<span class="op">!</span><span class="kw">dir.exists</span>(<span class="kw">file.path</span>(from,dir)))){<span class="kw">stop</span>(<span class="st">&quot;Invalid.&quot;</span>)}
    <span class="cf">if</span>(<span class="kw">any</span>(<span class="op">!</span><span class="kw">dir.exists</span>(<span class="kw">file.path</span>(to,dir)))){<span class="kw">stop</span>(<span class="st">&quot;Invalid.&quot;</span>)}
    pb &lt;-<span class="st"> </span>utils<span class="op">::</span><span class="kw">txtProgressBar</span>(<span class="dt">min=</span><span class="dv">0</span>,<span class="dt">max=</span><span class="dv">1</span>,<span class="dt">style=</span><span class="dv">3</span>)
    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(dir)){
        files0 &lt;-<span class="st"> </span><span class="kw">dir</span>(<span class="kw">file.path</span>(from,dir[i]))
        files1 &lt;-<span class="st"> </span><span class="kw">dir</span>(<span class="kw">file.path</span>(to,dir[i]))
        names &lt;-<span class="st"> </span>files0[<span class="op">!</span>files0 <span class="op">%in%</span><span class="st"> </span>files1]
        <span class="cf">for</span>(j <span class="cf">in</span> <span class="kw">seq_along</span>(names)){
            utils<span class="op">::</span><span class="kw">setTxtProgressBar</span>(<span class="dt">pb=</span>pb,<span class="dt">value=</span>(i<span class="op">-</span><span class="dv">1</span>)<span class="op">/</span><span class="dv">3</span><span class="op">+</span>(j<span class="op">*</span>i)<span class="op">/</span>(<span class="dv">3</span><span class="op">*</span><span class="kw">length</span>(names)))
            <span class="kw">file.copy</span>(<span class="dt">from=</span><span class="kw">file.path</span>(from,dir[i],names[j]),
                      <span class="dt">to=</span><span class="kw">file.path</span>(to,dir[i],names[j]),
                      <span class="dt">copy.date=</span><span class="ot">TRUE</span>)
        }
        utils<span class="op">::</span><span class="kw">setTxtProgressBar</span>(<span class="dt">pb=</span>pb,<span class="dt">value=</span>i<span class="op">/</span><span class="dv">3</span>)
    }
}

<span class="co"># update(from=&quot;results&quot;,to=&quot;//tsclient/N/palasso/results&quot;)</span>
    
}</code></pre></div>
</div>
<div id="preparation" class="section level1">
<h1>Preparation</h1>
<p>Last data download started on 2019-03-26 (R version 3.5.3).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### Downloading &quot;Isoform Expression Quantification&quot;. <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

directory &lt;-<span class="st"> </span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>)
<span class="kw">setwd</span>(directory)

<span class="co"># Retrieving cancer types:</span>
project &lt;-<span class="st"> </span>TCGAbiolinks<span class="op">::</span><span class="kw">getGDCprojects</span>()<span class="op">$</span>id
project &lt;-<span class="st"> </span>project[<span class="kw">grepl</span>(<span class="dt">x=</span>project,<span class="dt">pattern=</span><span class="st">&quot;TCGA&quot;</span>)]

<span class="co"># Downloading isoform expression data:</span>
y &lt;-<span class="st"> </span>X &lt;-<span class="st"> </span><span class="kw">list</span>()
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(project)){
    query &lt;-<span class="st"> </span>TCGAbiolinks<span class="op">::</span><span class="kw">GDCquery</span>(<span class="dt">project=</span>project[i],
                <span class="dt">data.category=</span><span class="st">&quot;Transcriptome Profiling&quot;</span>,
                <span class="dt">data.type=</span><span class="st">&quot;Isoform Expression Quantification&quot;</span>)
    TCGAbiolinks<span class="op">::</span><span class="kw">GDCdownload</span>(query,<span class="dt">method=</span><span class="st">&quot;api&quot;</span>,<span class="dt">directory=</span>directory)
    <span class="kw">trace</span>(TCGAbiolinks<span class="op">:::</span>readTranscriptomeProfiling,<span class="dt">tracer=</span><span class="kw">quote</span>(ignore.case&lt;-<span class="ot">TRUE</span>))
    X[[i]] &lt;-<span class="st"> </span>TCGAbiolinks<span class="op">::</span><span class="kw">GDCprepare</span>(query,<span class="dt">directory=</span>directory)
    X[[i]][,<span class="kw">c</span>(<span class="st">&quot;miRNA_ID&quot;</span>,<span class="st">&quot;reads_per_million_miRNA_mapped&quot;</span>,
              <span class="st">&quot;cross-mapped&quot;</span>,<span class="st">&quot;miRNA_region&quot;</span>)] &lt;-<span class="st"> </span><span class="ot">NULL</span>
    y[[i]] &lt;-<span class="st"> </span><span class="kw">rep</span>(project[i],<span class="dt">times=</span><span class="kw">length</span>(<span class="kw">unique</span>(X[[i]]<span class="op">$</span>barcode)))
}

<span class="kw">save</span>(<span class="dt">list=</span><span class="kw">c</span>(<span class="st">&quot;y&quot;</span>,<span class="st">&quot;X&quot;</span>),<span class="dt">file=</span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="st">&quot;isoform_raw.RData&quot;</span>))
<span class="kw">load</span>(<span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="st">&quot;isoform_raw.RData&quot;</span>),<span class="dt">verbose=</span><span class="ot">TRUE</span>)

<span class="co"># Merging isoform expression data:</span>
Xs &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="dt">what=</span>rbind,<span class="dt">args=</span>X) <span class="co"># sparse matrix</span>
y &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="dt">what=</span><span class="st">&quot;c&quot;</span>,<span class="dt">args=</span>y)

<span class="co"># Transform to matrix</span>
Xs<span class="op">$</span>isoform_coords &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern=</span><span class="st">&quot;hg38:chr&quot;</span>,<span class="dt">replacement=</span><span class="st">&quot;&quot;</span>,<span class="dt">x=</span>Xs<span class="op">$</span>isoform_coords)
samples &lt;-<span class="st"> </span><span class="kw">unique</span>(Xs<span class="op">$</span>barcode)
covariates &lt;-<span class="st"> </span><span class="kw">unique</span>(Xs<span class="op">$</span>isoform_coords)
row &lt;-<span class="st"> </span><span class="kw">match</span>(Xs<span class="op">$</span>barcode,samples)
col &lt;-<span class="st"> </span><span class="kw">match</span>(Xs<span class="op">$</span>isoform_coords,covariates)
X &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">sparseMatrix</span>(<span class="dt">i=</span>row,<span class="dt">j=</span>col,<span class="dt">x=</span>Xs<span class="op">$</span>read_count,<span class="dt">dimnames=</span><span class="kw">list</span>(samples,covariates))

<span class="co"># Order by molecular location</span>
split &lt;-<span class="st"> </span><span class="kw">strsplit</span>(<span class="dt">x=</span><span class="kw">colnames</span>(X),<span class="dt">split=</span><span class="st">&quot;:|-&quot;</span>)
chr &lt;-<span class="st"> </span><span class="kw">sapply</span>(split,<span class="cf">function</span>(x) x[[<span class="dv">1</span>]])
pos &lt;-<span class="st"> </span><span class="kw">sapply</span>(split,<span class="cf">function</span>(x) x[[<span class="dv">2</span>]])
order &lt;-<span class="st"> </span><span class="kw">order</span>(chr,pos)
X &lt;-<span class="st"> </span>X[,order]

<span class="cf">if</span>(<span class="ot">FALSE</span>){ <span class="co"># testing</span>
    i &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">seq_len</span>(<span class="kw">nrow</span>(Xs)),<span class="dt">size=</span><span class="dv">1</span>)
    Xs<span class="op">$</span>read_count[i]
    X[Xs<span class="op">$</span>barcode[i],Xs<span class="op">$</span>isoform_coords[i]]
}

<span class="kw">save</span>(<span class="dt">list=</span><span class="kw">c</span>(<span class="st">&quot;y&quot;</span>,<span class="st">&quot;X&quot;</span>),<span class="dt">file=</span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="st">&quot;isoform_all.RData&quot;</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### Downloading &quot;miRNA Expression Quantification&quot;. <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

directory &lt;-<span class="st"> </span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>)
<span class="kw">setwd</span>(directory)

<span class="co"># Downloading data.</span>
project &lt;-<span class="st"> </span>TCGAbiolinks<span class="op">::</span><span class="kw">getGDCprojects</span>()<span class="op">$</span>id 
project &lt;-<span class="st"> </span>project[<span class="kw">grepl</span>(<span class="dt">x=</span>project,<span class="dt">pattern=</span><span class="st">&quot;TCGA&quot;</span>)]
y &lt;-<span class="st"> </span>X &lt;-<span class="st"> </span><span class="kw">list</span>()
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(project)){
    query &lt;-<span class="st"> </span>TCGAbiolinks<span class="op">::</span><span class="kw">GDCquery</span>(<span class="dt">project=</span>project[i],
                           <span class="dt">data.category=</span><span class="st">&quot;Transcriptome Profiling&quot;</span>,
                           <span class="dt">data.type=</span><span class="st">&quot;miRNA Expression Quantification&quot;</span>)
    TCGAbiolinks<span class="op">::</span><span class="kw">GDCdownload</span>(query,<span class="dt">method=</span><span class="st">&quot;api&quot;</span>,<span class="dt">directory=</span>directory)
    <span class="kw">trace</span>(TCGAbiolinks<span class="op">:::</span>readTranscriptomeProfiling,<span class="dt">tracer=</span><span class="kw">quote</span>(ignore.case&lt;-<span class="ot">TRUE</span>))
    data &lt;-<span class="st"> </span>TCGAbiolinks<span class="op">::</span><span class="kw">GDCprepare</span>(query,<span class="dt">directory=</span>directory)
    X[[i]] &lt;-<span class="st"> </span><span class="kw">t</span>(data[,<span class="kw">c</span>(<span class="kw">seq</span>(<span class="dt">from=</span><span class="dv">2</span>,<span class="dt">to=</span><span class="kw">ncol</span>(data),<span class="dt">by=</span><span class="dv">3</span>))])
    y[[i]] &lt;-<span class="st"> </span><span class="kw">rep</span>(project[i],<span class="dt">times=</span><span class="kw">nrow</span>(X[[i]]))
}

<span class="kw">save</span>(<span class="dt">list=</span><span class="kw">c</span>(<span class="st">&quot;y&quot;</span>,<span class="st">&quot;X&quot;</span>),<span class="dt">file=</span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="st">&quot;miRNA_raw.RData&quot;</span>))
<span class="kw">load</span>(<span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="st">&quot;miRNA_raw.RData&quot;</span>))

X &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="dt">what=</span>rbind,<span class="dt">args=</span>X)
y &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="dt">what=</span><span class="st">&quot;c&quot;</span>,<span class="dt">args=</span>y)
<span class="kw">rownames</span>(X) &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern=</span><span class="st">&quot;read_count_&quot;</span>,<span class="dt">replacement=</span><span class="st">&quot;&quot;</span>,<span class="dt">x=</span><span class="kw">rownames</span>(X))

<span class="kw">save</span>(<span class="dt">list=</span><span class="kw">c</span>(<span class="st">&quot;y&quot;</span>,<span class="st">&quot;X&quot;</span>),<span class="dt">file=</span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="st">&quot;miRNA_all.RData&quot;</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### Downloading &quot;Gene Expression Quantification&quot;. <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

directory &lt;-<span class="st"> </span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>)
<span class="kw">setwd</span>(directory)

<span class="co"># Retrieving cancer types:</span>
project &lt;-<span class="st"> </span>TCGAbiolinks<span class="op">::</span><span class="kw">getGDCprojects</span>()<span class="op">$</span>id
project &lt;-<span class="st"> </span>project[<span class="kw">grepl</span>(<span class="dt">x=</span>project,<span class="dt">pattern=</span><span class="st">&quot;TCGA&quot;</span>)]

<span class="co"># Downloading data:</span>
<span class="kw">memory.limit</span>(<span class="dt">size=</span><span class="dv">16000</span>) <span class="co"># Activate virtual memory in system control!</span>
y &lt;-<span class="st"> </span>X &lt;-<span class="st"> </span><span class="kw">list</span>()
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(project)){
    query &lt;-<span class="st"> </span>TCGAbiolinks<span class="op">::</span><span class="kw">GDCquery</span>(<span class="dt">project=</span>project[i],
                <span class="dt">data.category=</span><span class="st">&quot;Transcriptome Profiling&quot;</span>,
                <span class="dt">data.type=</span><span class="st">&quot;Gene Expression Quantification&quot;</span>,
                <span class="dt">workflow.type=</span><span class="st">&quot;HTSeq - Counts&quot;</span>); <span class="kw">gc</span>()
    TCGAbiolinks<span class="op">::</span><span class="kw">GDCdownload</span>(<span class="dt">query=</span>query,<span class="dt">method=</span><span class="st">&quot;api&quot;</span>,<span class="dt">directory=</span>directory); <span class="kw">gc</span>()
    <span class="kw">trace</span>(TCGAbiolinks<span class="op">:::</span>readTranscriptomeProfiling,<span class="dt">tracer=</span><span class="kw">quote</span>(ignore.case&lt;-<span class="ot">TRUE</span>)); <span class="kw">gc</span>()
    X[[i]] &lt;-<span class="st"> </span>TCGAbiolinks<span class="op">::</span><span class="kw">GDCprepare</span>(query,<span class="dt">directory=</span>directory); <span class="kw">gc</span>()
    y[[i]] &lt;-<span class="st"> </span><span class="kw">rep</span>(project[i],<span class="dt">times=</span><span class="kw">ncol</span>(X[[i]])); <span class="kw">gc</span>()
}

<span class="kw">save</span>(<span class="dt">list=</span><span class="kw">c</span>(<span class="st">&quot;y&quot;</span>,<span class="st">&quot;X&quot;</span>),<span class="dt">file=</span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="st">&quot;gene_raw.RData&quot;</span>))
<span class="kw">load</span>(<span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="st">&quot;gene_raw.RData&quot;</span>))

genes &lt;-<span class="st"> </span>SummarizedExperiment<span class="op">::</span><span class="kw">rowData</span>(X[[<span class="dv">1</span>]])
mart &lt;-<span class="st"> </span>biomaRt<span class="op">::</span><span class="kw">useMart</span>(<span class="st">&quot;ensembl&quot;</span>,<span class="dt">dataset=</span><span class="st">&quot;hsapiens_gene_ensembl&quot;</span>) <span class="co"># </span>
char &lt;-<span class="st"> </span>biomaRt<span class="op">::</span><span class="kw">getBM</span>(<span class="dt">attributes=</span><span class="kw">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>,<span class="st">&quot;chromosome_name&quot;</span>,<span class="st">&quot;transcript_start&quot;</span>,<span class="st">&quot;gene_biotype&quot;</span>),<span class="dt">filters=</span><span class="kw">c</span>(<span class="st">&quot;biotype&quot;</span>,<span class="st">&quot;chromosome_name&quot;</span>),<span class="dt">values=</span><span class="kw">list</span>(<span class="st">&quot;protein_coding&quot;</span>,<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">22</span>,<span class="st">&quot;X&quot;</span>)),<span class="dt">mart=</span>mart)
select &lt;-<span class="st"> </span>genes<span class="op">$</span>ensembl_gene_id[genes<span class="op">$</span>ensembl_gene_id <span class="op">%in%</span><span class="st"> </span>char<span class="op">$</span>ensembl_gene_id]

X &lt;-<span class="st"> </span><span class="kw">lapply</span>(X,<span class="cf">function</span>(x) <span class="kw">t</span>(SummarizedExperiment<span class="op">::</span><span class="kw">assays</span>(x)<span class="op">$</span><span class="st">&quot;HTSeq - Counts&quot;</span>[select,]))
X &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="dt">what=</span>rbind,<span class="dt">args=</span>X)
y &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="dt">what=</span><span class="st">&quot;c&quot;</span>,<span class="dt">args=</span>y)

<span class="kw">save</span>(<span class="dt">list=</span><span class="kw">c</span>(<span class="st">&quot;y&quot;</span>,<span class="st">&quot;X&quot;</span>),<span class="dt">file=</span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="st">&quot;gene_all.RData&quot;</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### Downloading &quot;Copy Number Variation&quot;. <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

directory &lt;-<span class="st"> </span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>)
<span class="kw">setwd</span>(directory)

project &lt;-<span class="st"> </span>TCGAbiolinks<span class="op">::</span><span class="kw">getGDCprojects</span>()<span class="op">$</span>id
project &lt;-<span class="st"> </span>project[<span class="kw">grepl</span>(<span class="dt">x=</span>project,<span class="dt">pattern=</span><span class="st">&quot;TCGA&quot;</span>)]

y &lt;-<span class="st"> </span>X &lt;-<span class="st"> </span><span class="kw">list</span>()
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(project)){
    query &lt;-<span class="st"> </span>TCGAbiolinks<span class="op">::</span><span class="kw">GDCquery</span>(<span class="dt">project=</span>project[i],
                <span class="dt">data.category=</span><span class="st">&quot;Copy Number Variation&quot;</span>,
                <span class="dt">data.type=</span><span class="st">&quot;Masked Copy Number Segment&quot;</span>)
    TCGAbiolinks<span class="op">::</span><span class="kw">GDCdownload</span>(<span class="dt">query=</span>query,<span class="dt">method=</span><span class="st">&quot;api&quot;</span>,<span class="dt">directory=</span>directory)
    <span class="kw">trace</span>(TCGAbiolinks<span class="op">:::</span>readTranscriptomeProfiling,<span class="dt">tracer=</span><span class="kw">quote</span>(ignore.case&lt;-<span class="ot">TRUE</span>))
    X[[i]] &lt;-<span class="st"> </span>TCGAbiolinks<span class="op">::</span><span class="kw">GDCprepare</span>(query,<span class="dt">directory=</span>directory)
    y[[i]] &lt;-<span class="st"> </span><span class="kw">rep</span>(project[i],<span class="dt">times=</span><span class="kw">length</span>(<span class="kw">unique</span>(X[[i]]<span class="op">$</span>Sample))) <span class="co"># correct?</span>
}

<span class="kw">save</span>(<span class="dt">list=</span><span class="kw">c</span>(<span class="st">&quot;y&quot;</span>,<span class="st">&quot;X&quot;</span>),<span class="dt">file=</span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="st">&quot;CNV_raw.RData&quot;</span>))
<span class="kw">load</span>(<span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="st">&quot;CNV_raw.RData&quot;</span>),<span class="dt">verbose=</span><span class="ot">TRUE</span>)

<span class="co"># Merging CNV data:</span>
Xs &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="dt">what=</span>rbind,<span class="dt">args=</span>X) <span class="co"># sparse matrix</span>
y &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="dt">what=</span><span class="st">&quot;c&quot;</span>,<span class="dt">args=</span>y)
<span class="co">#table(Xs$Sample)</span>

<span class="co"># Prepare cutting.</span>
cut &lt;-<span class="st"> </span><span class="kw">list</span>()
cut<span class="op">$</span>chr &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">22</span>,<span class="st">&quot;X&quot;</span>)
cut<span class="op">$</span>start &lt;-<span class="st"> </span><span class="kw">sapply</span>(cut<span class="op">$</span>chr,<span class="cf">function</span>(x) <span class="kw">min</span>(Xs<span class="op">$</span>Start[Xs<span class="op">$</span>Chromosome<span class="op">==</span>x]))
cut<span class="op">$</span>end &lt;-<span class="st"> </span><span class="kw">sapply</span>(cut<span class="op">$</span>chr,<span class="cf">function</span>(x) <span class="kw">max</span>(Xs<span class="op">$</span>End[Xs<span class="op">$</span>Chromosome<span class="op">==</span>x]))
cut<span class="op">$</span>length &lt;-<span class="st"> </span>cut<span class="op">$</span>end<span class="op">-</span>cut<span class="op">$</span>start
cut<span class="op">$</span>dist &lt;-<span class="st"> </span><span class="kw">sum</span>(cut<span class="op">$</span>length)<span class="op">/</span><span class="dv">10000</span>
cut<span class="op">$</span>num &lt;-<span class="st"> </span><span class="kw">round</span>(cut<span class="op">$</span>length<span class="op">/</span>cut<span class="op">$</span>dist)

<span class="co"># Create covariates.</span>
cov &lt;-<span class="st"> </span><span class="kw">list</span>()
cov<span class="op">$</span>p &lt;-<span class="st"> </span><span class="kw">sum</span>(cut<span class="op">$</span>num)
cov<span class="op">$</span>chromosome &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">sapply</span>(cut<span class="op">$</span>chr,<span class="cf">function</span>(i) <span class="kw">rep</span>(i,<span class="dt">times=</span>cut<span class="op">$</span>num[i])))
cov<span class="op">$</span>location &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">sapply</span>(cut<span class="op">$</span>chr,<span class="cf">function</span>(i) <span class="kw">round</span>(<span class="kw">seq</span>(<span class="dt">from=</span>cut<span class="op">$</span>start[i],<span class="dt">to=</span>cut<span class="op">$</span>end[i],<span class="dt">length.out=</span>cut<span class="op">$</span>num[i]))))
cov<span class="op">$</span>name &lt;-<span class="st"> </span><span class="kw">paste0</span>(cov<span class="op">$</span>chromosome,<span class="st">&quot;:&quot;</span>,cov<span class="op">$</span>location)

<span class="co"># Create indices for each covariate.</span>
index &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="kw">list</span>(<span class="kw">integer</span>()),<span class="dt">times=</span>cov<span class="op">$</span>p)
pb &lt;-<span class="st"> </span>utils<span class="op">::</span><span class="kw">txtProgressBar</span>(<span class="dt">min=</span><span class="dv">0</span>,<span class="dt">max=</span>cov<span class="op">$</span>p,<span class="dt">style=</span><span class="dv">3</span>)
<span class="cf">for</span>(j <span class="cf">in</span> <span class="kw">seq_len</span>(cov<span class="op">$</span>p)){
    utils<span class="op">::</span><span class="kw">setTxtProgressBar</span>(<span class="dt">pb=</span>pb,<span class="dt">value=</span>j)
    index[[j]] &lt;-<span class="st"> </span><span class="kw">which</span>((Xs<span class="op">$</span>Chromosome<span class="op">==</span>cov<span class="op">$</span>chromosome[j]) <span class="op">&amp;</span><span class="st"> </span>
<span class="st">        </span>(Xs<span class="op">$</span>Start<span class="op">&lt;=</span>cov<span class="op">$</span>location[j]) <span class="op">&amp;</span><span class="st"> </span>(cov<span class="op">$</span>location[j]<span class="op">&lt;=</span>Xs<span class="op">$</span>End)) <span class="co"># consider &lt;</span>
}

<span class="co"># Expand indices to matrix.</span>
X &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>,<span class="dt">nrow=</span><span class="kw">length</span>(<span class="kw">unique</span>(Xs<span class="op">$</span>Sample)),<span class="dt">ncol=</span>cov<span class="op">$</span>p,
             <span class="dt">dimnames=</span><span class="kw">list</span>(<span class="kw">unique</span>(Xs<span class="op">$</span>Sample),cov<span class="op">$</span>name))
<span class="cf">for</span>(j <span class="cf">in</span> <span class="kw">seq_along</span>(index)){
    mean &lt;-<span class="st"> </span>Xs<span class="op">$</span>Segment_Mean[index[[j]]]
    i &lt;-<span class="st"> </span>Xs<span class="op">$</span>Sample[index[[j]]]
    X[i,j] &lt;-<span class="st"> </span>mean
}

<span class="cf">if</span>(<span class="ot">FALSE</span>){ <span class="co"># test</span>
    sample &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">rownames</span>(X),<span class="dt">size=</span><span class="dv">1</span>)
    covariate &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">colnames</span>(X),<span class="dt">size=</span><span class="dv">1</span>)
    split &lt;-<span class="st"> </span><span class="kw">strsplit</span>(covariate,<span class="dt">split=</span><span class="st">&quot;:&quot;</span>)[[<span class="dv">1</span>]]
    a &lt;-<span class="st"> </span>X[sample,covariate]
    b &lt;-<span class="st"> </span>Xs<span class="op">$</span>Segment_Mean[(Xs<span class="op">$</span>Sample<span class="op">==</span>sample) <span class="op">&amp;</span><span class="st"> </span>(Xs<span class="op">$</span>Chromosome<span class="op">==</span>split[<span class="dv">1</span>]) <span class="op">&amp;</span><span class="st"> </span>(Xs<span class="op">$</span>Start<span class="op">&lt;=</span><span class="kw">as.numeric</span>(split[<span class="dv">2</span>])) <span class="op">&amp;</span><span class="st"> </span>(<span class="kw">as.numeric</span>(split[<span class="dv">2</span>]) <span class="op">&lt;</span><span class="st"> </span>Xs<span class="op">$</span>End)]
    <span class="kw">all</span>(a<span class="op">==</span>b)
}

<span class="kw">save</span>(<span class="dt">list=</span><span class="kw">c</span>(<span class="st">&quot;y&quot;</span>,<span class="st">&quot;X&quot;</span>,<span class="st">&quot;index&quot;</span>),<span class="dt">file=</span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="st">&quot;CNV_all.RData&quot;</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### Extracting samples of interest. <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

type &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;isoform&quot;</span>,<span class="st">&quot;miRNA&quot;</span>,<span class="st">&quot;CNV&quot;</span>,<span class="st">&quot;gene&quot;</span>)

<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(type)){
    <span class="kw">cat</span>(type[i],<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)
    <span class="kw">load</span>(<span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="kw">paste0</span>(type[i],<span class="st">&quot;_all.RData&quot;</span>)),<span class="dt">verbose=</span><span class="ot">TRUE</span>)
    
    <span class="co"># TCGA barcode</span>
    barcode &lt;-<span class="st"> </span><span class="kw">rownames</span>(X)
    code &lt;-<span class="st"> </span><span class="kw">sapply</span>(barcode,<span class="cf">function</span>(x) <span class="kw">strsplit</span>(x,<span class="dt">split=</span><span class="st">&quot;-&quot;</span>))
    code &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">do.call</span>(<span class="dt">what=</span>rbind,<span class="dt">args=</span>code))
    <span class="kw">colnames</span>(code) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;project&quot;</span>,<span class="st">&quot;TSS&quot;</span>,<span class="st">&quot;participant&quot;</span>,<span class="st">&quot;sample_vial&quot;</span>,
                        <span class="st">&quot;portion_analyte&quot;</span>,<span class="st">&quot;plate&quot;</span>,<span class="st">&quot;center&quot;</span>)
    code<span class="op">$</span>sample &lt;-<span class="st"> </span><span class="kw">substr</span>(code<span class="op">$</span>sample_vial,<span class="dt">start=</span><span class="dv">1</span>,<span class="dt">stop=</span><span class="dv">2</span>)
    code<span class="op">$</span>vial &lt;-<span class="st"> </span><span class="kw">substr</span>(code<span class="op">$</span>sample_vial,<span class="dt">start=</span><span class="dv">3</span>,<span class="dt">stop=</span><span class="dv">3</span>)
    code<span class="op">$</span>portion &lt;-<span class="st"> </span><span class="kw">substr</span>(code<span class="op">$</span>portion_analyte,<span class="dt">start=</span><span class="dv">1</span>,<span class="dt">stop=</span><span class="dv">2</span>)
    code<span class="op">$</span>analyte &lt;-<span class="st"> </span><span class="kw">substr</span>(code<span class="op">$</span>portion_analyte,<span class="dt">start=</span><span class="dv">3</span>,<span class="dt">stop=</span><span class="dv">3</span>)
    code<span class="op">$</span>sample_vial &lt;-<span class="st"> </span>code<span class="op">$</span>portion_analyte &lt;-<span class="st"> </span><span class="ot">NULL</span>
    
    <span class="co"># solid tumour (except blood for LAML)</span>
    solid &lt;-<span class="st"> </span>(code<span class="op">$</span>sample<span class="op">==</span><span class="st">&quot;01&quot;</span> <span class="op">|</span><span class="st"> </span>(y<span class="op">==</span><span class="st">&quot;TCGA-LAML&quot;</span> <span class="op">&amp;</span><span class="st"> </span>code<span class="op">$</span>sample<span class="op">==</span><span class="st">&quot;03&quot;</span>))
    X &lt;-<span class="st"> </span>X[solid,]
    y &lt;-<span class="st"> </span>y[solid]

    <span class="co"># unique samples</span>
    unique &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">duplicated</span>(<span class="kw">substr</span>(<span class="kw">rownames</span>(X),<span class="dt">start=</span><span class="dv">1</span>,<span class="dt">stop=</span><span class="dv">12</span>))
    X &lt;-<span class="st"> </span>X[unique,]
    y &lt;-<span class="st"> </span>y[unique]
    
    <span class="kw">save</span>(<span class="dt">list=</span><span class="kw">c</span>(<span class="st">&quot;y&quot;</span>,<span class="st">&quot;X&quot;</span>),<span class="dt">file=</span><span class="kw">file.path</span>(path,<span class="st">&quot;data&quot;</span>,<span class="kw">paste0</span>(type[i],<span class="st">&quot;_sub.RData&quot;</span>)))
}

<span class="co"># isoform: n=9'794, p=194'595, k=32</span>
<span class="co"># miRNA: n=9'794, p=1'881, k=32</span>
<span class="co"># gene: n=9'830, p=19'602, k=33</span>
<span class="co"># CNV: n=10'578, p=10'000, k=33</span>

## Understanding barcodes:
<span class="co"># overview: https://wiki.nci.nih.gov/display/TCGA/TCGA+barcode</span>
<span class="co"># details: https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables</span>

## Understanding replicate samples:
<span class="co"># http://gdac.broadinstitute.org/runs/sampleReports/latest/READ_Replicate_Samples.html</span></code></pre></div>
</div>
<div id="analysis" class="section level1">
<h1>Analysis</h1>
<p>Last data analysis started on 2019-04-12 (R version 3.5.3).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### Analysing the TCGA data. <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

<span class="cf">for</span>(type <span class="cf">in</span> <span class="kw">c</span>(<span class="st">&quot;miRNA&quot;</span>,<span class="st">&quot;isoform&quot;</span>,<span class="st">&quot;CNV&quot;</span>,<span class="st">&quot;gene&quot;</span>)){ 
  <span class="kw">library</span>(Matrix)
  <span class="cf">for</span>(k <span class="cf">in</span> <span class="kw">c</span>(<span class="dv">207</span>,<span class="kw">sample</span>(<span class="dv">528</span>))){
    <span class="kw">set.seed</span>(k); <span class="kw">cat</span>(k,<span class="st">&quot; &quot;</span>)
    <span class="cf">if</span>(type <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;isoform&quot;</span>,<span class="st">&quot;miRNA&quot;</span>) <span class="op">&amp;</span><span class="st"> </span>k <span class="op">&gt;</span><span class="st"> </span><span class="dv">496</span>){<span class="cf">next</span>}
    <span class="cf">if</span>(type <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;CNV&quot;</span>,<span class="st">&quot;gene&quot;</span>) <span class="op">&amp;</span><span class="st"> </span>k <span class="op">&gt;</span><span class="st"> </span><span class="dv">528</span>){<span class="cf">next</span>}
    
    <span class="co"># searching for missing cancer-cancer combinations</span>
    <span class="kw">rm</span>(<span class="dt">list=</span><span class="kw">setdiff</span>(<span class="kw">ls</span>(),<span class="kw">c</span>(<span class="st">&quot;k&quot;</span>,<span class="st">&quot;type&quot;</span>,<span class="st">&quot;path&quot;</span>,<span class="st">&quot;save&quot;</span>,<span class="st">&quot;file.exists&quot;</span>,<span class="st">&quot;file.remove&quot;</span>))); <span class="kw">gc</span>()
    file0 &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;results/&quot;</span>,type,<span class="st">&quot;_start_&quot;</span>,k,<span class="st">&quot;.RData&quot;</span>)
    file1 &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;results/&quot;</span>,type,<span class="st">&quot;_loss_&quot;</span>,k,<span class="st">&quot;.RData&quot;</span>)
    <span class="cf">if</span>(<span class="kw">file.exists</span>(file0)<span class="op">||</span><span class="kw">file.exists</span>(file1)){<span class="cf">next</span>}
    <span class="kw">save</span>(<span class="dt">object=</span>k,<span class="dt">file=</span>file0)
    <span class="kw">load</span>(<span class="kw">paste0</span>(<span class="st">&quot;data/&quot;</span>,type,<span class="st">&quot;_sub.RData&quot;</span>))
    
    <span class="co"># indicating the cancer-cancer combination</span>
    cancer &lt;-<span class="st"> </span><span class="kw">substring</span>(<span class="dt">text=</span><span class="kw">unique</span>(y),<span class="dt">first=</span><span class="dv">6</span>)
    comb &lt;-<span class="st"> </span>utils<span class="op">::</span><span class="kw">combn</span>(<span class="dt">x=</span>cancer,<span class="dt">m=</span><span class="dv">2</span>)
    select &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;TCGA-&quot;</span>,comb[,k])
    y &lt;-<span class="st"> </span><span class="kw">ifelse</span>(y<span class="op">==</span>select[<span class="dv">1</span>],<span class="dv">1</span>,<span class="kw">ifelse</span>(y<span class="op">==</span>select[<span class="dv">2</span>],<span class="dv">0</span>,<span class="ot">NA</span>))
    <span class="kw">rm</span>(cancer,select)
    
    <span class="co"># removing other cancer types</span>
    cond &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(y)
    y &lt;-<span class="st"> </span>y[cond]
    X &lt;-<span class="st"> </span>X[cond,]
    <span class="kw">rm</span>(cond)
    
    <span class="co"># pre-processing</span>
    <span class="cf">if</span>(type <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;isoform&quot;</span>,<span class="st">&quot;miRNA&quot;</span>)){
      x &lt;-<span class="st"> </span>palasso<span class="op">:::</span><span class="kw">.prepare</span>(X,<span class="dt">cutoff=</span><span class="st">&quot;zero&quot;</span>)
    } <span class="cf">else</span> <span class="cf">if</span>(type<span class="op">==</span><span class="st">&quot;gene&quot;</span>){
      x &lt;-<span class="st"> </span>palasso<span class="op">:::</span><span class="kw">.prepare</span>(X,<span class="dt">cutoff=</span><span class="st">&quot;knee&quot;</span>)
    } <span class="cf">else</span> <span class="cf">if</span>(type<span class="op">==</span><span class="st">&quot;CNV&quot;</span>){
      x &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">X=</span>X,<span class="dt">Z=</span><span class="kw">sign</span>(X))
      x &lt;-<span class="st"> </span><span class="kw">lapply</span>(x,<span class="cf">function</span>(x) <span class="kw">scale</span>(x))
      <span class="kw">attributes</span>(x)<span class="op">$</span>info &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">n=</span><span class="kw">nrow</span>(X),<span class="dt">p=</span><span class="kw">ncol</span>(X),<span class="dt">prop=</span><span class="kw">mean</span>(x<span class="op">$</span>Z<span class="op">==</span><span class="dv">0</span>))
    }
    <span class="kw">rm</span>(X)
    
    <span class="co"># cross-validation</span>
    loss &lt;-<span class="st"> </span><span class="kw">tryCatch</span>(<span class="dt">expr=</span>palasso<span class="op">:::</span><span class="kw">.predict</span>(<span class="dt">y=</span>y,<span class="dt">X=</span>x,<span class="dt">nfolds.int=</span><span class="dv">10</span>),<span class="dt">error=</span><span class="cf">function</span>(e) palasso<span class="op">:::</span><span class="kw">.predict</span>(<span class="dt">y=</span>y,<span class="dt">X=</span>x,<span class="dt">nfolds.int=</span><span class="dv">10</span>))
    
    <span class="co"># information</span>
    loss<span class="op">$</span>info &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dt">k=</span>k,
                       <span class="dt">y0=</span>comb[<span class="dv">2</span>,k],
                       <span class="dt">y1=</span>comb[<span class="dv">1</span>,k],
                       <span class="dt">n0=</span><span class="kw">sum</span>(y<span class="op">==</span><span class="dv">0</span>),
                       <span class="dt">n1=</span><span class="kw">sum</span>(y<span class="op">==</span><span class="dv">1</span>),
                       <span class="kw">attributes</span>(x)<span class="op">$</span>info,
                       loss<span class="op">$</span>info)
    
    <span class="co"># refit</span>
    object &lt;-<span class="st"> </span>palasso<span class="op">::</span><span class="kw">palasso</span>(<span class="dt">y=</span>y,<span class="dt">X=</span>x,<span class="dt">nfolds=</span><span class="dv">10</span>,<span class="dt">family=</span><span class="st">&quot;binomial&quot;</span>,<span class="dt">standard=</span><span class="ot">TRUE</span>,<span class="dt">elastic=</span><span class="ot">TRUE</span>,<span class="dt">shrink=</span><span class="ot">TRUE</span>)
    
    model &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">names</span>(object),<span class="st">&quot;elastic&quot;</span>,
               <span class="kw">paste0</span>(<span class="st">&quot;paired.&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;adaptive&quot;</span>,<span class="st">&quot;standard&quot;</span>,<span class="st">&quot;combined&quot;</span>)))
    
    <span class="cf">for</span>(max <span class="cf">in</span> <span class="kw">c</span>(<span class="dv">10</span>,<span class="dv">50</span>,<span class="ot">Inf</span>)){
      temp &lt;-<span class="st"> </span><span class="kw">list</span>()
      temp<span class="op">$</span>nzero &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">model=</span>model,<span class="dt">x=</span><span class="ot">NA</span>,<span class="dt">z=</span><span class="ot">NA</span>)
      <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(model)){
        coef &lt;-<span class="st"> </span>palasso<span class="op">:::</span><span class="kw">coef.palasso</span>(<span class="dt">object=</span>object,<span class="dt">model=</span>model[i],<span class="dt">max=</span>max)
        temp<span class="op">$</span>nzero<span class="op">$</span>x[i] &lt;-<span class="st"> </span><span class="kw">sum</span>(coef<span class="op">$</span>x<span class="op">!=</span><span class="dv">0</span>)
        temp<span class="op">$</span>nzero<span class="op">$</span>z[i] &lt;-<span class="st"> </span><span class="kw">sum</span>(coef<span class="op">$</span>z<span class="op">!=</span><span class="dv">0</span>)
      }
      temp<span class="op">$</span>select &lt;-<span class="st"> </span>palasso<span class="op">:::</span><span class="kw">subset.palasso</span>(<span class="dt">x=</span>object,<span class="dt">max=</span>max,
                                              <span class="dt">model=</span><span class="st">&quot;paired.adaptive&quot;</span>)<span class="op">$</span>palasso<span class="op">$</span>select
      temp<span class="op">$</span>weights &lt;-<span class="st"> </span>palasso<span class="op">:::</span><span class="kw">weights.palasso</span>(<span class="dt">object=</span>object,<span class="dt">max=</span>max,
                                                <span class="dt">model=</span><span class="st">&quot;paired.adaptive&quot;</span>)
      temp<span class="op">$</span>coef &lt;-<span class="st"> </span>palasso<span class="op">:::</span><span class="kw">coef.palasso</span>(<span class="dt">object=</span>object,<span class="dt">max=</span>max,
                                          <span class="dt">model=</span><span class="st">&quot;paired.adaptive&quot;</span>)
      loss[[<span class="kw">paste0</span>(<span class="st">&quot;fit&quot;</span>,max)]] &lt;-<span class="st"> </span>temp
    }
    
    <span class="kw">save</span>(<span class="dt">object=</span>loss,<span class="dt">file=</span>file1)
    <span class="kw">file.remove</span>(file0)
  }
  
  index &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="kw">grepl</span>(<span class="kw">dir</span>(),<span class="dt">pattern=</span><span class="st">&quot;sessionInfo&quot;</span>))
  <span class="kw">sink</span>(<span class="kw">paste0</span>(<span class="st">&quot;sessionInfo&quot;</span>,index<span class="op">+</span><span class="dv">1</span>,<span class="st">&quot;.txt&quot;</span>))
  <span class="kw">date</span>()
  utils<span class="op">::</span><span class="kw">sessionInfo</span>()
  devtools<span class="op">::</span><span class="kw">session_info</span>()
  <span class="kw">sink</span>()
  
}</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The function &lt;&lt;collect&gt;&gt; loads all files from PATH including PATTERN in the file name, loads OBJECT into a list, and executes a function call.</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

<span class="co"># path: folder</span>
<span class="co"># pattern: character, or NULL (all files)</span>
<span class="co"># object: character vector, or NULL (all objects)</span>
<span class="co"># what: function call</span>

collect &lt;-<span class="st"> </span><span class="cf">function</span>(<span class="dt">path=</span><span class="kw">getwd</span>(),<span class="dt">pattern=</span><span class="st">&quot;&quot;</span>,<span class="dt">object=</span><span class="ot">NULL</span>,<span class="dt">what=</span><span class="st">&quot;rbind&quot;</span>){
    OBJECT =<span class="st"> </span>object
    <span class="co"># identify files</span>
    files &lt;-<span class="st"> </span><span class="kw">dir</span>(path)
    files &lt;-<span class="st"> </span>files[<span class="kw">grepl</span>(<span class="dt">x=</span>files,<span class="dt">pattern=</span>pattern)]
    files &lt;-<span class="st"> </span>files[<span class="kw">grepl</span>(<span class="dt">x=</span>files,<span class="dt">pattern=</span><span class="st">&quot;.RData&quot;</span>)]
    number &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern=</span><span class="kw">paste0</span>(pattern,<span class="st">&quot;|.RData&quot;</span>),<span class="dt">replacement=</span><span class="st">&quot;&quot;</span>,<span class="dt">x=</span>files)
    files &lt;-<span class="st"> </span>files[<span class="kw">order</span>(<span class="kw">as.numeric</span>(number))] <span class="co"># trial1</span>
    names &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern=</span><span class="st">&quot;.RData&quot;</span>,<span class="dt">replacement=</span><span class="st">&quot;&quot;</span>,<span class="dt">x=</span>files) <span class="co"># trial1</span>
    <span class="cf">if</span>(<span class="kw">length</span>(files)<span class="op">==</span><span class="dv">0</span>){<span class="kw">stop</span>(<span class="st">&quot;Invalid datasets.&quot;</span>)}
    <span class="co"># load data</span>
    all &lt;-<span class="st"> </span><span class="kw">list</span>()
    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(files)){
        x &lt;-<span class="st"> </span><span class="kw">load</span>(<span class="kw">file.path</span>(path,files[i]))
        x &lt;-<span class="st"> </span><span class="kw">eval</span>(<span class="kw">parse</span>(<span class="dt">text=</span>x))
        <span class="cf">if</span>(<span class="kw">is.null</span>(OBJECT)){
            all[[i]] &lt;-<span class="st"> </span>x
            <span class="kw">names</span>(all)[i] &lt;-<span class="st"> </span>names[i] 
        } <span class="cf">else</span> {
            <span class="cf">for</span>(j <span class="cf">in</span> <span class="kw">seq_along</span>(OBJECT)){
                all[[OBJECT[j]]][[i]] &lt;-<span class="st"> </span>x[[OBJECT[j]]]
                <span class="kw">names</span>(all[[OBJECT[j]]])[i] &lt;-<span class="st"> </span>names[i]
            }
        }
    }
    <span class="co"># fuse data</span>
    <span class="cf">if</span>(<span class="kw">is.null</span>(OBJECT)){
        all &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="dt">what=</span>what,<span class="dt">args=</span>all)
    } <span class="cf">else</span> {
        all &lt;-<span class="st"> </span><span class="kw">lapply</span>(all,<span class="cf">function</span>(x) <span class="kw">do.call</span>(<span class="dt">what=</span>what,<span class="dt">args=</span>x))
    }
    <span class="kw">return</span>(all)
}

LOSS &lt;-<span class="st"> </span><span class="kw">list</span>()
type &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;gene&quot;</span>,<span class="st">&quot;isoform&quot;</span>,<span class="st">&quot;miRNA&quot;</span>,<span class="st">&quot;CNV&quot;</span>)
<span class="co">#type &lt;- &quot;miRNA&quot;</span>
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(type)){
  LOSS[[type[i]]] &lt;-<span class="st"> </span><span class="kw">collect</span>(<span class="dt">path=</span><span class="st">&quot;results&quot;</span>,
                             <span class="dt">pattern=</span><span class="kw">paste0</span>(type[i],<span class="st">&quot;_loss_&quot;</span>),
                             <span class="dt">object=</span><span class="kw">c</span>(<span class="st">&quot;deviance&quot;</span>,<span class="st">&quot;auc&quot;</span>,<span class="st">&quot;class&quot;</span>,<span class="st">&quot;info&quot;</span>,
                                      <span class="kw">paste0</span>(<span class="st">&quot;fit&quot;</span>,<span class="kw">c</span>(<span class="dv">10</span>,<span class="dv">50</span>,<span class="ot">Inf</span>))))
}
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(LOSS)){
  <span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>){
    <span class="kw">colnames</span>(LOSS[[i]][[j]])[<span class="kw">colnames</span>(LOSS[[i]][[j]])<span class="op">==</span><span class="st">&quot;paired.adaptive&quot;</span>] &lt;-<span class="st"> &quot;paired&quot;</span>
  }
}

<span class="co">#type &lt;- &quot;gene&quot;</span>
<span class="co">#a &lt;- LOSS[[type]]$deviance[rownames(LOSS[[type]]$deviance)==&quot;10&quot;,&quot;paired&quot;]</span>
<span class="co">#b &lt;- LOSS[[type]]$deviance[rownames(LOSS[[type]]$deviance)==&quot;10&quot;,&quot;elastic&quot;]</span>
<span class="co">#mean(a&lt;b)</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### Testing for significant differences. <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>
<span class="co">#&lt;&lt;collect&gt;&gt;</span>

row &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;gene&quot;</span>,<span class="st">&quot;isoform&quot;</span>,<span class="st">&quot;miRNA&quot;</span>,<span class="st">&quot;CNV&quot;</span>)
col &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;10&quot;</span>,<span class="st">&quot;Inf&quot;</span>)
lay &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;standard_x&quot;</span>,<span class="st">&quot;standard_z&quot;</span>,<span class="st">&quot;standard_xz&quot;</span>,
         <span class="st">&quot;adaptive_x&quot;</span>,<span class="st">&quot;adaptive_z&quot;</span>,<span class="st">&quot;adaptive_xz&quot;</span>,
         <span class="st">&quot;elastic&quot;</span>) <span class="co"># added &quot;elastic&quot;</span>

M &lt;-<span class="st"> </span><span class="kw">array</span>(<span class="ot">NA</span>,<span class="dt">dim=</span><span class="kw">c</span>(<span class="kw">length</span>(row),<span class="kw">length</span>(col),<span class="kw">length</span>(lay)),<span class="dt">dimnames=</span><span class="kw">list</span>(row,col,lay))

<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(row)){
  
  loss &lt;-<span class="st"> </span>LOSS[[row[i]]][<span class="kw">c</span>(<span class="st">&quot;info&quot;</span>,<span class="st">&quot;deviance&quot;</span>)]
  
  y0 &lt;-<span class="st"> </span><span class="kw">as.character</span>(loss<span class="op">$</span>info<span class="op">$</span>y0)
  y1 &lt;-<span class="st"> </span><span class="kw">as.character</span>(loss<span class="op">$</span>info<span class="op">$</span>y1)
  cancer &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">unique</span>(<span class="kw">c</span>(y0,y1)))
  Z &lt;-<span class="st"> </span>palasso<span class="op">:::</span><span class="kw">.design</span>(<span class="dt">x=</span>cancer)
  
  <span class="cf">for</span>(j <span class="cf">in</span> <span class="kw">seq_along</span>(col)){
    
    <span class="co"># differences</span>
    cond &lt;-<span class="st"> </span><span class="kw">rownames</span>(loss<span class="op">$</span>deviance)<span class="op">==</span>col[j]
    
    <span class="cf">for</span>(k <span class="cf">in</span> <span class="kw">seq_along</span>(lay)){
      
      fill &lt;-<span class="st"> </span>loss<span class="op">$</span>deviance[cond,lay[k]] <span class="op">-</span><span class="st"> </span>loss<span class="op">$</span>deviance[cond,<span class="st">&quot;paired&quot;</span>]
      X &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>,<span class="dt">nrow=</span><span class="kw">length</span>(cancer),<span class="dt">ncol=</span><span class="kw">length</span>(cancer),
                  <span class="dt">dimnames=</span><span class="kw">list</span>(cancer,cancer))
      X[<span class="kw">cbind</span>(y0,y1)] &lt;-<span class="st"> </span>X[<span class="kw">cbind</span>(y1,y0)] &lt;-<span class="st"> </span>fill
      X[<span class="kw">lower.tri</span>(X)] &lt;-<span class="st"> </span><span class="ot">NA</span>
      
      <span class="co"># p-values</span>
      pvalue &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="ot">NA</span>,<span class="dt">times=</span><span class="kw">max</span>(Z))
      <span class="cf">for</span>(l <span class="cf">in</span> <span class="kw">seq_len</span>(<span class="kw">max</span>(Z))){
        x &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(X[Z<span class="op">==</span>l])
        <span class="cf">if</span>(col[j]<span class="op">==</span><span class="st">&quot;10&quot;</span>){
          alternative &lt;-<span class="st"> &quot;greater&quot;</span> <span class="co"># Never use &quot;two.sided&quot;!</span>
        }
        <span class="cf">if</span>(col[j]<span class="op">==</span><span class="st">&quot;Inf&quot;</span>){
          alternative &lt;-<span class="st"> &quot;less&quot;</span> <span class="co"># Never use &quot;two.sided&quot;!</span>
        }
        pvalue[l] &lt;-<span class="st"> </span>stats<span class="op">::</span><span class="kw">wilcox.test</span>(<span class="dt">x=</span>x,<span class="dt">alternative=</span>alternative,
                                 <span class="dt">exact=</span><span class="ot">FALSE</span>)<span class="op">$</span>p.value
      }
      
      <span class="co"># Simes</span>
      M[i,j,k] &lt;-<span class="st"> </span>palasso<span class="op">:::</span><span class="kw">.combine</span>(pvalue,<span class="dt">method=</span><span class="st">&quot;simes&quot;</span>)
    }
  }
}

<span class="co"># Table SIG: significance</span>
constraint &lt;-<span class="st"> &quot;10&quot;</span>
table &lt;-<span class="st"> </span><span class="kw">format</span>(M[,constraint,<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>],<span class="dt">digits=</span><span class="dv">1</span>,<span class="dt">scientific=</span><span class="ot">FALSE</span>)
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_len</span>(<span class="kw">nrow</span>(table))){
    <span class="cf">for</span>(j <span class="cf">in</span> <span class="kw">seq_len</span>(<span class="kw">ncol</span>(table))){
        <span class="cf">if</span>(M[i,constraint,j]<span class="op">&gt;=</span><span class="fl">0.05</span>){
            table[i,j] &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;{</span><span class="ch">\\</span><span class="st">color{gray}{&quot;</span>,table[i,j],<span class="st">&quot;}}&quot;</span>)
        }
    }
}
one &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;&quot;</span>,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">text{standard}&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">text{adaptive}&quot;</span>,<span class="st">&quot;&quot;</span>)
two &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">text{&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>,<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>),<span class="st">&quot;}&quot;</span>)
<span class="kw">rownames</span>(table) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">text{&quot;</span>,<span class="kw">rownames</span>(table),<span class="st">&quot;}&quot;</span>)
table &lt;-<span class="st"> </span><span class="kw">rbind</span>(one,two,table,<span class="dt">deparse.level=</span><span class="dv">0</span>)
<span class="kw">rownames</span>(table)[<span class="dv">1</span>] &lt;-<span class="st"> &quot;~&quot;</span>
xtable &lt;-<span class="st"> </span>xtable<span class="op">::</span><span class="kw">xtable</span>(table,<span class="dt">align=</span><span class="kw">c</span>(<span class="st">&quot;r&quot;</span>,<span class="st">&quot;|&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;|&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;c&quot;</span>))
xtable<span class="op">::</span><span class="kw">print.xtable</span>(xtable,<span class="dt">type=</span><span class="st">&quot;latex&quot;</span>,<span class="dt">include.colnames=</span><span class="ot">FALSE</span>,<span class="dt">sanitize.text.function=</span>identity)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### Comparison with elastic net. <span class="al">###</span>
<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>
<span class="co">#&lt;&lt;collect&gt;&gt;</span>

row &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;gene&quot;</span>,<span class="st">&quot;isoform&quot;</span>,<span class="st">&quot;miRNA&quot;</span>,<span class="st">&quot;CNV&quot;</span>)
col &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;10&quot;</span>,<span class="st">&quot;50&quot;</span>,<span class="st">&quot;Inf&quot;</span>)
better &lt;-<span class="st"> </span>worse &lt;-<span class="st"> </span>less &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>,<span class="dt">nrow=</span><span class="kw">length</span>(row),<span class="dt">ncol=</span><span class="kw">length</span>(col),
                 <span class="dt">dimnames=</span><span class="kw">list</span>(row,col))
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(row)){
  <span class="cf">for</span>(j <span class="cf">in</span> <span class="kw">seq_along</span>(col)){
    <span class="co"># proportion of improvements (cross-validation)</span>
    cond &lt;-<span class="st"> </span><span class="kw">rownames</span>(LOSS[[row[i]]]<span class="op">$</span>deviance)<span class="op">==</span>col[j]
    loss &lt;-<span class="st"> </span>LOSS[[row[i]]]<span class="op">$</span>deviance[cond,<span class="kw">c</span>(<span class="st">&quot;paired&quot;</span>,<span class="st">&quot;elastic&quot;</span>)]
    better[i,j] &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="kw">mean</span>(loss[,<span class="st">&quot;paired&quot;</span>]<span class="op">&lt;</span>loss[,<span class="st">&quot;elastic&quot;</span>]),<span class="dt">digits=</span><span class="dv">2</span>)
    worse[i,j] &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="kw">mean</span>(loss[,<span class="st">&quot;paired&quot;</span>]<span class="op">&gt;</span>loss[,<span class="st">&quot;elastic&quot;</span>]),<span class="dt">digits=</span><span class="dv">2</span>)
    <span class="co"># average difference in nzero (refitted models)</span>
    df_paired &lt;-<span class="st"> </span><span class="kw">apply</span>(LOSS[[row[i]]][[<span class="kw">paste0</span>(<span class="st">&quot;fit&quot;</span>,col[j])]],<span class="dv">1</span>,<span class="cf">function</span>(x)   <span class="kw">sum</span>(x<span class="op">$</span>nzero[x<span class="op">$</span>nzero[,<span class="st">&quot;model&quot;</span>]<span class="op">==</span><span class="st">&quot;paired.adaptive&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>)]))
    df_elastic &lt;-<span class="st"> </span><span class="kw">apply</span>(LOSS[[row[i]]][[<span class="kw">paste0</span>(<span class="st">&quot;fit&quot;</span>,col[j])]],<span class="dv">1</span>,<span class="cf">function</span>(x)   <span class="kw">sum</span>(x<span class="op">$</span>nzero[x<span class="op">$</span>nzero[,<span class="st">&quot;model&quot;</span>]<span class="op">==</span><span class="st">&quot;elastic95&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>)]))
    df_diff &lt;-<span class="st"> </span>df_elastic<span class="op">-</span>df_paired
    less[i,j] &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="kw">mean</span>(df_diff),<span class="dt">digits=</span><span class="dv">2</span>)
    <span class="co">#graphics::hist(df_diff,main=paste(row[i],col[j]),xlim=c(-1,1)*max(abs(df_diff)))</span>
  }
}
better
worse
less</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### Analysing the refitted models. <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>
<span class="co">#&lt;&lt;collect&gt;&gt;</span>

<span class="co"># Table SEL: selected model</span>
nzero &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;fit&quot;</span>,<span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">10</span>,<span class="ot">Inf</span>))
model &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">paste0</span>(<span class="st">&quot;standard_&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>)),
           <span class="kw">paste0</span>(<span class="st">&quot;adaptive_&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>)),
           <span class="st">&quot;between_xz&quot;</span>,<span class="st">&quot;within_xz&quot;</span>)
type &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;gene&quot;</span>,<span class="st">&quot;isoform&quot;</span>,<span class="st">&quot;miRNA&quot;</span>,<span class="st">&quot;CNV&quot;</span>)
table &lt;-<span class="st"> </span><span class="kw">array</span>(<span class="ot">NA</span>,<span class="dt">dim=</span><span class="kw">c</span>(<span class="kw">length</span>(nzero),<span class="kw">length</span>(model),<span class="kw">length</span>(type)),
               <span class="dt">dimnames=</span><span class="kw">list</span>(nzero,model,type))
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(nzero)){
    <span class="cf">for</span>(j <span class="cf">in</span> <span class="kw">seq_along</span>(model)){
        <span class="cf">for</span>(k <span class="cf">in</span> <span class="kw">seq_along</span>(type)){
            sub &lt;-<span class="st"> </span>LOSS[[type[k]]][[nzero[i]]]
            table[i,j,k] &lt;-<span class="st"> </span><span class="kw">sum</span>(sub[,<span class="st">&quot;select&quot;</span>]<span class="op">==</span>model[j])
        }
    }
}
<span class="kw">colSums</span>(table[<span class="st">&quot;fit10&quot;</span>,,]) <span class="co"># CHECK WHETHER COMPLETE!</span>
table &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="kw">prop.table</span>(table[<span class="st">&quot;fit10&quot;</span>,,],<span class="dt">margin=</span><span class="dv">2</span>),<span class="dt">digits=</span><span class="dv">2</span>)
table &lt;-<span class="st"> </span><span class="kw">t</span>(table)
table &lt;-<span class="st"> </span>table[,<span class="kw">apply</span>(table,<span class="dv">2</span>,<span class="cf">function</span>(x) <span class="kw">any</span>(x<span class="op">!=</span><span class="dv">0</span>))]
<span class="kw">rownames</span>(table) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">text{&quot;</span>,<span class="kw">rownames</span>(table),<span class="st">&quot;}&quot;</span>)
xtable &lt;-<span class="st"> </span>xtable<span class="op">::</span><span class="kw">xtable</span>(table,<span class="dt">align=</span><span class="kw">c</span>(<span class="st">&quot;r&quot;</span>,<span class="st">&quot;|&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;c&quot;</span>))
xtable<span class="op">::</span><span class="kw">print.xtable</span>(xtable,<span class="dt">type=</span><span class="st">&quot;latex&quot;</span>,<span class="dt">include.colnames=</span><span class="ot">FALSE</span>,<span class="dt">sanitize.text.function=</span>identity)

<span class="co"># selected weights and covariates</span>
type &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;gene&quot;</span>,<span class="st">&quot;isoform&quot;</span>,<span class="st">&quot;miRNA&quot;</span>,<span class="st">&quot;CNV&quot;</span>)
group &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>)
model &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">paste0</span>(<span class="st">&quot;standard_&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>)),
           <span class="kw">paste0</span>(<span class="st">&quot;adaptive_&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>)),
           <span class="st">&quot;paired.adaptive&quot;</span>,<span class="st">&quot;elastic&quot;</span>) <span class="co"># added &quot;elastic&quot; , paste0(&quot;elastic&quot;,c(100,75,50,25))</span>
weights10 &lt;-<span class="st"> </span>weightsInf &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>,<span class="dt">nrow=</span><span class="kw">length</span>(group),<span class="dt">ncol=</span><span class="kw">length</span>(type),
                <span class="dt">dimnames=</span><span class="kw">list</span>(group,type))
coef10 &lt;-<span class="st"> </span>coefInf &lt;-<span class="st"> </span><span class="kw">array</span>(<span class="ot">NA</span>,<span class="dt">dim=</span><span class="kw">c</span>(<span class="kw">length</span>(group),<span class="kw">length</span>(type),<span class="kw">length</span>(model)),
               <span class="dt">dimnames=</span><span class="kw">list</span>(group,type,model))
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(group)){
    <span class="cf">for</span>(j <span class="cf">in</span> <span class="kw">seq_along</span>(type)){
        weights10[,j] &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(<span class="kw">sapply</span>(LOSS[[type[j]]]<span class="op">$</span>fit10[,<span class="st">&quot;weights&quot;</span>],colMeans))
        weightsInf[,j] &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(<span class="kw">sapply</span>(LOSS[[type[j]]]<span class="op">$</span>fitInf[,<span class="st">&quot;weights&quot;</span>],colMeans))
        <span class="cf">for</span>(k <span class="cf">in</span> <span class="kw">seq_along</span>(model)){
            coef10[i,j,k] &lt;-<span class="st"> </span><span class="kw">mean</span>(<span class="kw">sapply</span>(LOSS[[type[j]]]<span class="op">$</span>fit10[,<span class="st">&quot;nzero&quot;</span>], <span class="cf">function</span>(x) <span class="kw">sum</span>(x[x<span class="op">$</span>model<span class="op">==</span>model[k],group[i]])))
            coefInf[i,j,k] &lt;-<span class="st"> </span><span class="kw">mean</span>(<span class="kw">sapply</span>(LOSS[[type[j]]]<span class="op">$</span>fitInf[,<span class="st">&quot;nzero&quot;</span>], <span class="cf">function</span>(x) <span class="kw">sum</span>(x[x<span class="op">$</span>model<span class="op">==</span>model[k],group[i]])))
        }
    }
}
<span class="co"># coef10[&quot;x&quot;,,]+coef10[&quot;z&quot;,,]</span>


<span class="co"># with sparsity constraint</span>
<span class="kw">round</span>(<span class="kw">prop.table</span>(weights10,<span class="dt">margin=</span><span class="dv">2</span>),<span class="dv">2</span>)
<span class="kw">round</span>(<span class="kw">prop.table</span>(coef10[,,<span class="st">&quot;paired.adaptive&quot;</span>],<span class="dt">margin=</span><span class="dv">2</span>),<span class="dv">2</span>)
<span class="kw">round</span>(<span class="kw">colSums</span>(coef10[,,<span class="st">&quot;paired.adaptive&quot;</span>]),<span class="dv">2</span>)

<span class="co"># natural sparsity</span>
<span class="kw">round</span>(<span class="kw">prop.table</span>(weightsInf,<span class="dt">margin=</span><span class="dv">2</span>),<span class="dv">2</span>)
<span class="kw">round</span>(<span class="kw">prop.table</span>(coefInf[,,<span class="st">&quot;paired.adaptive&quot;</span>],<span class="dt">margin=</span><span class="dv">2</span>),<span class="dv">2</span>)
<span class="kw">round</span>(<span class="kw">colSums</span>(coefInf[,,<span class="st">&quot;paired.adaptive&quot;</span>]),<span class="dv">2</span>)
<span class="kw">round</span>(<span class="kw">colSums</span>(coefInf[,,<span class="st">&quot;elastic&quot;</span>])<span class="op">/</span><span class="kw">colSums</span>(coefInf[,,<span class="st">&quot;paired.adaptive&quot;</span>]),<span class="dv">1</span>) <span class="co"># multiple nzero of elastic net and paired lasso</span>

<span class="co"># Table NSC: number of non-zero coefficients</span>
table &lt;-<span class="st"> </span><span class="kw">round</span>(coefInf[<span class="st">&quot;x&quot;</span>,,]<span class="op">+</span>coefInf[<span class="st">&quot;z&quot;</span>,,])
<span class="kw">colnames</span>(table)[<span class="dv">7</span>] &lt;-<span class="st"> &quot;paired&quot;</span>
one &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;&quot;</span>,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">text{standard}&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">text{adaptive}&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">text{paired}&quot;</span>,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">text{elastic}&quot;</span>)
two &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">text{&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>,<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>,<span class="st">&quot;xz&quot;</span>,<span class="st">&quot;xz&quot;</span>),<span class="st">&quot;}&quot;</span>)
<span class="kw">rownames</span>(table) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">text{&quot;</span>,<span class="kw">rownames</span>(table),<span class="st">&quot;}&quot;</span>)
table &lt;-<span class="st"> </span><span class="kw">rbind</span>(one,two,table,<span class="dt">deparse.level=</span><span class="dv">0</span>)
<span class="kw">rownames</span>(table)[<span class="dv">1</span>] &lt;-<span class="st"> &quot;~&quot;</span>
xtable &lt;-<span class="st"> </span>xtable<span class="op">::</span><span class="kw">xtable</span>(table,<span class="dt">align=</span><span class="kw">c</span>(<span class="st">&quot;r&quot;</span>,<span class="st">&quot;|&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;|&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;|&quot;</span>,<span class="st">&quot;c&quot;</span>,<span class="st">&quot;|&quot;</span>,<span class="st">&quot;c&quot;</span>))
xtable<span class="op">::</span><span class="kw">print.xtable</span>(xtable,<span class="dt">type=</span><span class="st">&quot;latex&quot;</span>,<span class="dt">include.colnames=</span><span class="ot">FALSE</span>,<span class="dt">sanitize.text.function=</span>identity)</code></pre></div>
</div>
<div id="figures" class="section level1">
<h1>Figures</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### FIGURE CSW <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

<span class="kw">set.seed</span>(<span class="dv">1</span>)
overfit &lt;-<span class="st"> </span><span class="ot">TRUE</span>

<span class="co"># simulate</span>
n &lt;-<span class="st"> </span><span class="dv">10</span>
cx &lt;-<span class="st"> </span>stats<span class="op">::</span><span class="kw">rbeta</span>(<span class="dt">n=</span>n,<span class="dt">shape1=</span><span class="fl">0.9</span>,<span class="dt">shape2=</span><span class="dv">1</span>)
cz &lt;-<span class="st"> </span>stats<span class="op">::</span><span class="kw">rbeta</span>(<span class="dt">n=</span>n,<span class="dt">shape1=</span><span class="fl">0.4</span>,<span class="dt">shape2=</span><span class="fl">0.9</span>)

<span class="co"># collection</span>
x &lt;-<span class="st"> </span><span class="kw">list</span>()
y &lt;-<span class="st"> </span><span class="kw">list</span>()

<span class="co"># adaptive weights (X only)</span>
x[[<span class="dv">1</span>]] &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">1</span>,<span class="dt">times=</span>n)
<span class="cf">if</span>(overfit){x[[<span class="dv">1</span>]] &lt;-<span class="st"> </span>cx}
y[[<span class="dv">1</span>]] &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>,<span class="dt">times=</span>n)

<span class="co"># adaptive weights (Z only)</span>
x[[<span class="dv">2</span>]] &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>,<span class="dt">times=</span>n)
y[[<span class="dv">2</span>]] &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">1</span>,<span class="dt">times=</span>n)
<span class="cf">if</span>(overfit){y[[<span class="dv">2</span>]] &lt;-<span class="st"> </span>cz}

<span class="co"># adaptive weights (X and Z)</span>
x[[<span class="dv">3</span>]] &lt;-<span class="st"> </span>y[[<span class="dv">3</span>]] &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="fl">0.5</span>,<span class="dt">times=</span>n)
<span class="cf">if</span>(overfit){x[[<span class="dv">3</span>]] &lt;-<span class="st"> </span>cx}
<span class="cf">if</span>(overfit){y[[<span class="dv">3</span>]] &lt;-<span class="st"> </span>cz}

<span class="co"># within-pair weights</span>
x[[<span class="dv">4</span>]] &lt;-<span class="st"> </span>cx<span class="op">^</span><span class="dv">2</span><span class="op">/</span>(cx<span class="op">+</span>cz)
y[[<span class="dv">4</span>]] &lt;-<span class="st"> </span>cz<span class="op">^</span><span class="dv">2</span><span class="op">/</span>(cx<span class="op">+</span>cz)

<span class="co"># visualisation</span>
graphics<span class="op">::</span><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>),<span class="dt">mar=</span><span class="kw">c</span>(<span class="fl">4.5</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>),<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">0</span>))
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_len</span>(<span class="dv">4</span>)){
    palasso<span class="op">:::</span><span class="kw">plot_pairs</span>(<span class="dt">x=</span>x[[i]],<span class="dt">y=</span>y[[i]],<span class="dt">lwd=</span><span class="dv">4</span>)
    <span class="cf">if</span>(i<span class="op">==</span><span class="dv">1</span>){
        graphics<span class="op">::</span><span class="kw">mtext</span>(<span class="dt">text=</span><span class="st">&quot;covariate pair&quot;</span>,<span class="dt">side=</span><span class="dv">2</span>,<span class="dt">line=</span><span class="dv">1</span>)
    }
}</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### FIGURE DIA <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

ellipse &lt;-<span class="st"> </span><span class="cf">function</span>(x,y,<span class="dt">a=</span><span class="fl">0.2</span>,<span class="dt">b=</span><span class="fl">0.25</span>,<span class="dt">border=</span><span class="ot">NA</span>){
    n &lt;-<span class="st"> </span><span class="kw">max</span>(<span class="kw">c</span>(<span class="kw">length</span>(x),<span class="kw">length</span>(y)))
    <span class="cf">if</span>(<span class="kw">length</span>(x)<span class="op">==</span><span class="dv">1</span>){x &lt;-<span class="st"> </span><span class="kw">rep</span>(x,<span class="dt">times=</span>n)}
    <span class="cf">if</span>(<span class="kw">length</span>(y)<span class="op">==</span><span class="dv">1</span>){y &lt;-<span class="st"> </span><span class="kw">rep</span>(y,<span class="dt">times=</span>n)}
    <span class="cf">if</span>(<span class="kw">length</span>(border)<span class="op">==</span><span class="dv">1</span>){border &lt;-<span class="st"> </span><span class="kw">rep</span>(border,<span class="dt">times=</span>n)}
    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_len</span>(n)){
        angle &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dt">from=</span><span class="dv">0</span>,<span class="dt">to=</span><span class="dv">2</span><span class="op">*</span>pi,<span class="dt">length=</span><span class="dv">100</span>)
        xs &lt;-<span class="st"> </span>x[i] <span class="op">+</span><span class="st"> </span>a <span class="op">*</span><span class="st"> </span><span class="kw">cos</span>(angle)
        ys &lt;-<span class="st"> </span>y[i] <span class="op">+</span><span class="st"> </span>b <span class="op">*</span><span class="st"> </span><span class="kw">sin</span>(angle)
        graphics<span class="op">::</span><span class="kw">polygon</span>(<span class="dt">x=</span>xs,<span class="dt">y=</span>ys,<span class="dt">col=</span><span class="kw">grey</span>(<span class="fl">0.9</span>),<span class="dt">border=</span>border[i])
    }
}

cancer &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ACC&quot;</span>,<span class="st">&quot;BLCA&quot;</span>,<span class="st">&quot;BRCA&quot;</span>,<span class="st">&quot;UVM&quot;</span>)
number &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">80</span>,<span class="dv">409</span>,<span class="dv">1078</span>,<span class="dv">80</span>)
col &lt;-<span class="st"> </span>grDevices<span class="op">::</span><span class="kw">rgb</span>(<span class="dt">red=</span><span class="dv">0</span>,<span class="dt">green=</span><span class="dv">0</span>,<span class="dt">blue=</span><span class="kw">sample</span>(<span class="kw">seq</span>(<span class="dt">from=</span><span class="dv">75</span>,<span class="dt">to=</span><span class="dv">255</span>,<span class="dt">length.out=</span><span class="kw">length</span>(number))),<span class="dt">maxColorValue=</span><span class="dv">255</span>)

lwd &lt;-<span class="st"> </span><span class="kw">log</span>(<span class="dv">2</span><span class="op">*</span>number)<span class="op">-</span><span class="dv">2</span>
lwd =<span class="st"> </span><span class="kw">pmax</span>(<span class="dv">5</span><span class="op">*</span>number<span class="op">/</span><span class="kw">max</span>(number),<span class="dv">1</span>)
x1 &lt;-<span class="st"> </span><span class="fl">1.3</span>; x2 &lt;-<span class="st"> </span><span class="dv">2</span>; x3 &lt;-<span class="st"> </span><span class="dv">3</span>

<span class="kw">set.seed</span>(<span class="dv">1</span>)

<span class="co"># first layer (bubble)</span>
graphics<span class="op">::</span><span class="kw">plot.new</span>()
graphics<span class="op">::</span><span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))
graphics<span class="op">::</span><span class="kw">plot.window</span>(<span class="dt">xlim=</span><span class="kw">c</span>(<span class="fl">1.1</span>,<span class="fl">3.3</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="fl">1.6</span>,<span class="fl">1.6</span>))
<span class="kw">ellipse</span>(<span class="dt">x=</span>x1,<span class="dt">y=</span><span class="dv">0</span>)
graphics<span class="op">::</span><span class="kw">text</span>(<span class="dt">x=</span>x1,<span class="dt">y=</span><span class="dv">0</span>,<span class="dt">labels=</span><span class="st">&quot;TCGA&quot;</span>,<span class="dt">font=</span><span class="dv">2</span>,<span class="dt">adj=</span><span class="kw">c</span>(<span class="fl">0.5</span>,<span class="dv">0</span>))
graphics<span class="op">::</span><span class="kw">text</span>(<span class="dt">x=</span>x1,<span class="dt">y=</span><span class="dv">0</span>,<span class="dt">labels=</span><span class="st">&quot;n=9794&quot;</span>,<span class="dt">adj=</span><span class="kw">c</span>(<span class="fl">0.5</span>,<span class="fl">1.2</span>),<span class="dt">cex=</span><span class="fl">0.9</span>)

<span class="co"># second layer (colon)</span>
y1 &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dt">from=</span><span class="dv">1</span>,<span class="dt">to=</span><span class="op">-</span><span class="dv">1</span>,<span class="dt">length.out=</span><span class="kw">length</span>(cancer)<span class="op">+</span><span class="dv">1</span>)
graphics<span class="op">::</span><span class="kw">text</span>(<span class="dt">x=</span>x2,<span class="dt">y=</span>y1[<span class="kw">length</span>(cancer)],<span class="dt">labels=</span><span class="st">&quot;...&quot;</span>,<span class="dt">font=</span><span class="dv">2</span>,<span class="dt">srt=</span><span class="dv">90</span>)
y1 &lt;-<span class="st"> </span>y1[<span class="op">-</span><span class="kw">length</span>(cancer)]

<span class="co"># first-second layer (connect)</span>
graphics<span class="op">::</span><span class="kw">segments</span>(<span class="dt">x0=</span>x1<span class="op">+</span><span class="fl">0.2</span>,<span class="dt">y0=</span><span class="dv">0</span>,<span class="dt">x1=</span>x2<span class="op">-</span><span class="fl">0.2</span>,<span class="dt">y1=</span>y1,<span class="dt">lwd=</span>lwd,<span class="dt">col=</span>col)

<span class="co"># second layer (bubble)</span>
<span class="kw">ellipse</span>(<span class="dt">x=</span>x2,<span class="dt">y=</span>y1,<span class="dt">a=</span><span class="fl">0.2</span>,<span class="dt">b=</span><span class="fl">0.2</span>)
graphics<span class="op">::</span><span class="kw">text</span>(<span class="dt">x=</span>x2,<span class="dt">y=</span>y1,<span class="dt">labels=</span>cancer,<span class="dt">font=</span><span class="dv">2</span>,<span class="dt">col=</span>col,<span class="dt">adj=</span><span class="kw">c</span>(<span class="fl">0.5</span>,<span class="dv">0</span>))
graphics<span class="op">::</span><span class="kw">text</span>(<span class="dt">x=</span>x2,<span class="dt">y=</span>y1,<span class="dt">labels=</span><span class="kw">paste0</span>(<span class="st">&quot;n=&quot;</span>,number,<span class="st">&quot;&quot;</span>),<span class="dt">adj=</span><span class="kw">c</span>(<span class="fl">0.5</span>,<span class="fl">1.2</span>),<span class="dt">cex=</span><span class="fl">0.9</span>)

comb &lt;-<span class="st"> </span>utils<span class="op">::</span><span class="kw">combn</span>(<span class="dt">x=</span><span class="kw">seq_along</span>(y1),<span class="dt">m=</span><span class="dv">2</span>)

<span class="co"># third layer (colon)</span>
y2 &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dt">from=</span><span class="fl">1.5</span>,<span class="dt">to=</span><span class="op">-</span><span class="fl">1.5</span>,<span class="dt">length.out=</span><span class="kw">ncol</span>(comb)<span class="op">+</span><span class="dv">1</span>)
graphics<span class="op">::</span><span class="kw">text</span>(<span class="dt">x=</span>x3,<span class="dt">y=</span>y2[<span class="kw">ncol</span>(comb)],<span class="dt">labels=</span><span class="st">&quot;...&quot;</span>,<span class="dt">font=</span><span class="dv">2</span>,<span class="dt">srt=</span><span class="dv">90</span>)
y2 &lt;-<span class="st"> </span>y2[<span class="op">-</span><span class="kw">ncol</span>(comb)]

<span class="co"># second-third layer (connect)</span>
graphics<span class="op">::</span><span class="kw">segments</span>(<span class="dt">x0=</span><span class="fl">2.2</span>,<span class="dt">y0=</span>y1[comb[<span class="dv">1</span>,]],<span class="dt">x1=</span><span class="fl">2.7</span>,<span class="dt">y1=</span>y2,<span class="dt">lwd=</span>lwd[comb[<span class="dv">1</span>,]],<span class="dt">col=</span>col[comb[<span class="dv">1</span>,]])
graphics<span class="op">::</span><span class="kw">segments</span>(<span class="dt">x0=</span><span class="fl">2.2</span>,<span class="dt">y0=</span>y1[comb[<span class="dv">2</span>,]],<span class="dt">x1=</span><span class="fl">2.7</span>,<span class="dt">y1=</span>y2,<span class="dt">lwd=</span>lwd[comb[<span class="dv">2</span>,]],<span class="dt">col=</span>col[comb[<span class="dv">2</span>,]])

<span class="co"># third layer (bubble)</span>
<span class="kw">ellipse</span>(<span class="dt">x=</span>x3,<span class="dt">y=</span>y2,<span class="dt">a=</span><span class="fl">0.3</span>,<span class="dt">b=</span><span class="fl">0.22</span>)
graphics<span class="op">::</span><span class="kw">text</span>(<span class="dt">x=</span>x3,<span class="dt">y=</span>y2,<span class="dt">labels=</span><span class="kw">paste0</span>(cancer[comb[<span class="dv">1</span>,]],<span class="st">&quot; &quot;</span>),
               <span class="dt">font=</span><span class="dv">2</span>,<span class="dt">col=</span>col[comb[<span class="dv">1</span>,]],<span class="dt">adj=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">0</span>))
graphics<span class="op">::</span><span class="kw">text</span>(<span class="dt">x=</span>x3,<span class="dt">y=</span>y2,<span class="dt">labels=</span><span class="kw">paste0</span>(<span class="st">&quot; &quot;</span>,cancer[comb[<span class="dv">2</span>,]]),
               <span class="dt">font=</span><span class="dv">2</span>,<span class="dt">col=</span>col[comb[<span class="dv">2</span>,]],<span class="dt">adj=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>))
graphics<span class="op">::</span><span class="kw">text</span>(<span class="dt">x=</span>x3,<span class="dt">y=</span>y2,<span class="dt">labels=</span><span class="st">&quot;:&quot;</span>,<span class="dt">font=</span><span class="dv">2</span>,<span class="dt">adj=</span><span class="kw">c</span>(<span class="fl">0.5</span>,<span class="dv">0</span>))
labels &lt;-<span class="st"> </span><span class="kw">apply</span>(comb,<span class="dv">2</span>,<span class="cf">function</span>(x) <span class="kw">sum</span>(number[x]))
labels &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;n=&quot;</span>,labels,<span class="st">&quot;&quot;</span>)
graphics<span class="op">::</span><span class="kw">text</span>(<span class="dt">x=</span>x3,<span class="dt">y=</span>y2,<span class="dt">labels=</span>labels,<span class="dt">adj=</span><span class="kw">c</span>(<span class="fl">0.5</span>,<span class="fl">1.2</span>),<span class="dt">cex=</span><span class="fl">0.9</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### FIGURE CLA <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

graphics<span class="op">::</span><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">2</span>),<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>),<span class="dt">mar=</span><span class="kw">c</span>(<span class="fl">2.1</span>,<span class="fl">3.5</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>))

<span class="cf">for</span>(type <span class="cf">in</span> <span class="kw">c</span>(<span class="st">&quot;gene&quot;</span>,<span class="st">&quot;isoform&quot;</span>,<span class="st">&quot;miRNA&quot;</span>,<span class="st">&quot;CNV&quot;</span>)){
    
    loss &lt;-<span class="st"> </span>LOSS[[type]][<span class="kw">c</span>(<span class="st">&quot;deviance&quot;</span>,<span class="st">&quot;auc&quot;</span>,<span class="st">&quot;class&quot;</span>)]
    choice &lt;-<span class="st"> &quot;paired&quot;</span>
    loss &lt;-<span class="st"> </span><span class="kw">lapply</span>(loss,<span class="cf">function</span>(x) x[,<span class="kw">c</span>(<span class="kw">paste0</span>(<span class="st">&quot;standard_&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>)),<span class="kw">paste0</span>(<span class="st">&quot;adaptive_&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>)),choice)])
    
    <span class="cf">for</span>(constraint <span class="cf">in</span> <span class="kw">c</span>(<span class="st">&quot;10&quot;</span>)){ <span class="co"># c(&quot;5&quot;,&quot;10&quot;,&quot;Inf&quot;)</span>
        <span class="co"># change</span>
        sub &lt;-<span class="st"> </span><span class="kw">lapply</span>(loss,<span class="cf">function</span>(x) x[<span class="kw">rownames</span>(x)<span class="op">==</span>constraint,])
        palasso<span class="op">:::</span><span class="kw">plot_score</span>(sub<span class="op">$</span>deviance,<span class="dt">choice=</span>choice)
        change &lt;-<span class="st"> </span>sub<span class="op">$</span>deviance[,<span class="dv">7</span>]<span class="op">-</span>sub<span class="op">$</span>deviance[,<span class="op">-</span><span class="dv">7</span>]
        palasso<span class="op">:::</span><span class="kw">plot_box</span>(change,<span class="dt">ylab=</span><span class="st">&quot;change&quot;</span>,<span class="dt">zero=</span><span class="ot">TRUE</span>,<span class="dt">choice=</span><span class="ot">NA</span>)
        <span class="co"># info</span>
        info &lt;-<span class="st"> </span><span class="kw">list</span>()
        info<span class="op">$</span>select &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">which.min</span>(<span class="kw">apply</span>(sub<span class="op">$</span>deviance,<span class="dv">2</span>,median)[<span class="op">-</span><span class="dv">7</span>]))
        info<span class="op">$</span>DEV_paired &lt;-<span class="st"> </span><span class="kw">median</span>(sub<span class="op">$</span>deviance[,choice])
        info<span class="op">$</span>DEV_select &lt;-<span class="st"> </span><span class="kw">median</span>(sub<span class="op">$</span>deviance[,info<span class="op">$</span>select])
        info<span class="op">$</span>improve &lt;-<span class="st"> </span><span class="kw">mean</span>(sub<span class="op">$</span>deviance[,info<span class="op">$</span>select]<span class="op">&gt;</span>sub<span class="op">$</span>deviance[,choice])
        info<span class="op">$</span>AUC_paired &lt;-<span class="st"> </span><span class="kw">median</span>(sub<span class="op">$</span>auc[,choice])
        info<span class="op">$</span>CLASS_paired &lt;-<span class="st"> </span><span class="kw">median</span>(sub<span class="op">$</span>class[,choice])
        <span class="kw">print</span>(<span class="kw">as.data.frame</span>(info)) <span class="co"># important</span>
    }
}</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### FIGURE DEC <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

<span class="co">#graphics::par(oma=c(1.0,1.0,0,0),mar=c(1.5,3.0,0.5,0.5),mfrow=c(1,1))</span>
graphics<span class="op">::</span><span class="kw">par</span>(<span class="dt">oma=</span><span class="kw">c</span>(<span class="fl">1.0</span>,<span class="fl">1.0</span>,<span class="dv">0</span>,<span class="dv">0</span>),<span class="dt">mar=</span><span class="kw">c</span>(<span class="fl">1.5</span>,<span class="fl">3.0</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>),<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))

<span class="cf">for</span>(type <span class="cf">in</span> <span class="kw">c</span>(<span class="st">&quot;gene&quot;</span>,<span class="st">&quot;isoform&quot;</span>,<span class="st">&quot;miRNA&quot;</span>,<span class="st">&quot;CNV&quot;</span>)){
    models &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">paste0</span>(<span class="st">&quot;standard_&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>)),
                <span class="kw">paste0</span>(<span class="st">&quot;adaptive_&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>)),<span class="st">&quot;paired&quot;</span>)
    constraint &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;3&quot;</span>,<span class="st">&quot;4&quot;</span>,<span class="st">&quot;5&quot;</span>,<span class="st">&quot;10&quot;</span>,<span class="st">&quot;15&quot;</span>,<span class="st">&quot;20&quot;</span>,<span class="st">&quot;25&quot;</span>,<span class="st">&quot;50&quot;</span>,<span class="st">&quot;Inf&quot;</span>)
    
    loss &lt;-<span class="st"> </span>LOSS[[type]][<span class="st">&quot;deviance&quot;</span>]
    
    loss &lt;-<span class="st"> </span><span class="kw">lapply</span>(loss,<span class="cf">function</span>(x) x[,models])
    
    table &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>,<span class="dt">nrow=</span><span class="kw">length</span>(constraint),<span class="dt">ncol=</span><span class="kw">length</span>(models),
                    <span class="dt">dimnames=</span><span class="kw">list</span>(constraint,models))
    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(constraint)){
        sub &lt;-<span class="st"> </span><span class="kw">lapply</span>(loss,<span class="cf">function</span>(x) x[<span class="kw">rownames</span>(x)<span class="op">==</span>constraint[i],])
        table[i,] &lt;-<span class="st"> </span><span class="kw">apply</span>(sub<span class="op">$</span>deviance,<span class="dv">2</span>,median)
    }
    
    <span class="co"># table &lt;- log(table)</span>
    graphics<span class="op">::</span><span class="kw">plot.new</span>()
    graphics<span class="op">::</span><span class="kw">plot.window</span>(<span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="kw">length</span>(constraint)),<span class="dt">ylim=</span><span class="kw">range</span>(table))
    graphics<span class="op">::</span><span class="kw">box</span>()
    constraint[constraint<span class="op">==</span><span class="st">&quot;Inf&quot;</span>] &lt;-<span class="st"> &quot;n&quot;</span>
    graphics<span class="op">::</span><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">2</span>)
    graphics<span class="op">::</span><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>,<span class="dt">at=</span><span class="kw">seq_along</span>(constraint),<span class="dt">labels=</span>constraint,<span class="dt">tick=</span><span class="ot">FALSE</span>,<span class="dt">line=</span><span class="op">-</span><span class="dv">1</span>)

    <span class="cf">for</span>(k <span class="cf">in</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)){
        <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(models)){
            lty &lt;-<span class="st"> </span><span class="kw">ifelse</span>(i<span class="op">%in%</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>),<span class="dv">3</span>,<span class="kw">ifelse</span>(i<span class="op">%in%</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>),<span class="dv">2</span>,<span class="dv">1</span>))
            col &lt;-<span class="st"> </span><span class="kw">ifelse</span>(i<span class="op">==</span><span class="dv">7</span>,<span class="st">&quot;#00007F&quot;</span>,<span class="st">&quot;#FF3535&quot;</span>) 
            pch &lt;-<span class="st"> </span><span class="kw">ifelse</span>(i<span class="op">%in%</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>),<span class="st">&quot;x&quot;</span>,<span class="kw">ifelse</span>(i<span class="op">%in%</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">5</span>),<span class="st">&quot;z&quot;</span>,<span class="dv">1</span>))
            <span class="cf">if</span>(k<span class="op">==</span><span class="dv">1</span>){
                graphics<span class="op">::</span><span class="kw">lines</span>(table[,i],<span class="dt">col=</span>col,<span class="dt">lty=</span>lty,<span class="dt">lwd=</span><span class="dv">2</span>)
                graphics<span class="op">::</span><span class="kw">points</span>(table[,i],<span class="dt">col=</span><span class="st">&quot;white&quot;</span>,<span class="dt">pch=</span><span class="dv">16</span>,<span class="dt">cex=</span><span class="fl">1.2</span>)
            } <span class="cf">else</span> {
                graphics<span class="op">::</span><span class="kw">points</span>(table[,i],<span class="dt">col=</span>col,<span class="dt">pch=</span><span class="dv">1</span>,<span class="dt">font=</span><span class="dv">2</span>)  
            }
        }
    }
}
graphics<span class="op">::</span><span class="kw">title</span>(<span class="dt">ylab=</span><span class="st">&quot;deviance&quot;</span>,<span class="dt">line=</span><span class="fl">0.0</span>,<span class="dt">outer=</span><span class="ot">TRUE</span>)
graphics<span class="op">::</span><span class="kw">title</span>(<span class="dt">xlab=</span><span class="st">&quot;sparsity constraint&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;deviance&quot;</span>,<span class="dt">line=</span><span class="fl">0.0</span>,<span class="dt">outer=</span><span class="ot">TRUE</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### FIGURE CNV <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

graphics<span class="op">::</span><span class="kw">par</span>(<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>),<span class="dt">mar=</span><span class="kw">c</span>(<span class="fl">2.1</span>,<span class="fl">3.5</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>))
graphics<span class="op">::</span><span class="kw">layout</span>(<span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>),<span class="dt">nrow=</span><span class="dv">1</span>))

loss &lt;-<span class="st"> </span>LOSS[[type]][<span class="kw">c</span>(<span class="st">&quot;deviance&quot;</span>,<span class="st">&quot;auc&quot;</span>,<span class="st">&quot;class&quot;</span>)]
loss &lt;-<span class="st"> </span><span class="kw">lapply</span>(loss,<span class="cf">function</span>(x) x[<span class="kw">rownames</span>(x)<span class="op">==</span><span class="st">&quot;10&quot;</span>,])
model &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">paste0</span>(<span class="st">&quot;standard_&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>)),
           <span class="kw">paste0</span>(<span class="st">&quot;adaptive_&quot;</span>,<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>,<span class="st">&quot;z&quot;</span>,<span class="st">&quot;xz&quot;</span>)))

diff &lt;-<span class="st"> </span>loss<span class="op">$</span>auc[,<span class="st">&quot;paired&quot;</span>]<span class="op">-</span>loss<span class="op">$</span>auc[,model]
palasso<span class="op">:::</span><span class="kw">plot_box</span>(diff,<span class="dt">zero=</span><span class="ot">TRUE</span>,<span class="dt">invert=</span><span class="ot">TRUE</span>,<span class="dt">ylab=</span><span class="st">&quot;change&quot;</span>)
diff &lt;-<span class="st"> </span>loss<span class="op">$</span>class[,<span class="st">&quot;paired&quot;</span>]<span class="op">-</span>loss<span class="op">$</span>class[,model]
palasso<span class="op">:::</span><span class="kw">plot_box</span>(diff,<span class="dt">zero=</span><span class="ot">TRUE</span>,<span class="dt">ylab=</span><span class="st">&quot;change&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### FIGURE MAP <span class="al">###</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

loss &lt;-<span class="st"> </span>LOSS[[<span class="st">&quot;CNV&quot;</span>]][<span class="kw">c</span>(<span class="st">&quot;info&quot;</span>,<span class="st">&quot;auc&quot;</span>)]

cancer &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">unique</span>(<span class="kw">c</span>(<span class="kw">levels</span>(loss<span class="op">$</span>info<span class="op">$</span>y0),<span class="kw">levels</span>(loss<span class="op">$</span>info<span class="op">$</span>y1))))
X &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>,<span class="dt">nrow=</span><span class="kw">length</span>(cancer),<span class="dt">ncol=</span><span class="kw">length</span>(cancer),<span class="dt">dimnames=</span><span class="kw">list</span>(cancer,cancer))
<span class="co">#Z &lt;- palasso:::.design(x=cancer)</span>
y0 &lt;-<span class="st"> </span><span class="kw">as.character</span>(loss<span class="op">$</span>info<span class="op">$</span>y0)
y1 &lt;-<span class="st"> </span><span class="kw">as.character</span>(loss<span class="op">$</span>info<span class="op">$</span>y1)
X[<span class="kw">cbind</span>(y0,y1)] &lt;-<span class="st"> </span>X[<span class="kw">cbind</span>(y1,y0)] &lt;-<span class="st"> </span>loss<span class="op">$</span>auc[<span class="kw">rownames</span>(loss<span class="op">$</span>auc)<span class="op">==</span><span class="st">&quot;10&quot;</span>,<span class="st">&quot;paired&quot;</span>]

graphics<span class="op">::</span><span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="fl">0.5</span>,<span class="fl">3.0</span>,<span class="fl">3.0</span>,<span class="fl">0.5</span>))
<span class="kw">dimnames</span>(X) &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">dimnames</span>(X),<span class="cf">function</span>(x) <span class="kw">paste0</span>(<span class="st">&quot;  &quot;</span>,x,<span class="st">&quot;  &quot;</span>))
palasso<span class="op">:::</span><span class="kw">plot_table</span>(<span class="dt">X=</span>X,<span class="dt">margin=</span><span class="op">-</span><span class="dv">1</span>,<span class="dt">labels=</span><span class="ot">FALSE</span>,<span class="dt">las=</span><span class="dv">2</span>,<span class="dt">cex=</span><span class="fl">0.7</span>)
<span class="co">#sort(rowMeans(X,na.rm=TRUE),decreasing=TRUE)[1:2] # keep!</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### FIGURE COM <span class="al">###</span>

<span class="co"># 32 cancer types for isoform and miRNA</span>
<span class="co"># 33 cancer types for gene and CNV</span>

<span class="co">#rm(list=ls())</span>
<span class="co">#&lt;&lt;functions&gt;&gt;</span>

<span class="cf">for</span>(type <span class="cf">in</span> <span class="kw">c</span>(<span class="st">&quot;miRNA&quot;</span>)){
    
cancer &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">unique</span>(<span class="kw">as.character</span>(<span class="kw">unlist</span>(LOSS[[type]]<span class="op">$</span>info[,<span class="kw">c</span>(<span class="st">&quot;y0&quot;</span>,<span class="st">&quot;y1&quot;</span>)]))))

n &lt;-<span class="st"> </span><span class="kw">length</span>(cancer)
z &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(palasso<span class="op">:::</span><span class="kw">.design</span>(<span class="dt">x=</span>n))
x &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="kw">seq_len</span>(n),<span class="dt">each=</span>n)
y &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="kw">seq</span>(<span class="dt">from=</span>n,<span class="dt">to=</span><span class="dv">1</span>,<span class="dt">by=</span><span class="op">-</span><span class="dv">1</span>),<span class="dt">times=</span>n)
    
pch &lt;-<span class="st"> </span>z
pch[pch<span class="op">==</span><span class="dv">0</span>] &lt;-<span class="st"> </span><span class="ot">NA</span>
pex &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;.&quot;</span>,<span class="st">&quot;O&quot;</span>,<span class="st">&quot;*&quot;</span>,<span class="st">&quot;+&quot;</span>,<span class="st">&quot;o&quot;</span>,<span class="st">&quot;-&quot;</span>,<span class="st">&quot;'&quot;</span>,<span class="st">&quot;x&quot;</span>)

<span class="co"># colour</span>
base &lt;-<span class="st"> </span>grDevices<span class="op">::</span><span class="kw">colorRampPalette</span>(<span class="dt">colors=</span><span class="kw">c</span>(<span class="st">'darkblue'</span>,<span class="st">'blue'</span>,<span class="st">'red'</span>,<span class="st">'darkred'</span>))(n)
        
col &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="ot">NA</span>,<span class="dt">times=</span><span class="kw">length</span>(z))
col[z<span class="op">==</span><span class="dv">0</span>] &lt;-<span class="st"> &quot;white&quot;</span>
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_len</span>(n)){
    col[z<span class="op">==</span>i] &lt;-<span class="st"> </span>base[i]
}

graphics<span class="op">::</span><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>),<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>))
graphics<span class="op">::</span><span class="kw">plot.new</span>()
graphics<span class="op">::</span><span class="kw">plot.window</span>(<span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">1</span>,n),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">1</span>,n))
graphics<span class="op">::</span><span class="kw">points</span>(<span class="dt">x=</span>x[pch<span class="op">&lt;=</span><span class="dv">25</span>],<span class="dt">y=</span>y[pch<span class="op">&lt;=</span><span class="dv">25</span>],
                 <span class="dt">pch=</span>pch[pch<span class="op">&lt;=</span><span class="dv">25</span>],<span class="dt">col=</span>col[pch<span class="op">&lt;=</span><span class="dv">25</span>],<span class="dt">cex=</span><span class="fl">0.9</span>)
graphics<span class="op">::</span><span class="kw">points</span>(<span class="dt">x=</span>x[pch<span class="op">&gt;</span><span class="dv">25</span>],<span class="dt">y=</span>y[pch<span class="op">&gt;</span><span class="dv">25</span>],
                 <span class="dt">pch=</span>pex[(pch<span class="op">-</span><span class="dv">25</span>)[pch<span class="op">&gt;</span><span class="dv">25</span>]],<span class="dt">col=</span>col[pch<span class="op">&gt;</span><span class="dv">25</span>],<span class="dt">cex=</span><span class="fl">0.9</span>)
graphics<span class="op">::</span><span class="kw">segments</span>(<span class="dt">x0=</span><span class="dv">0</span>,<span class="dt">x1=</span>n<span class="op">+</span><span class="dv">1</span>,<span class="dt">y0=</span>n<span class="op">+</span><span class="dv">1</span>)
graphics<span class="op">::</span><span class="kw">segments</span>(<span class="dt">x0=</span>n<span class="op">+</span><span class="dv">1</span>,<span class="dt">y0=</span>n<span class="op">+</span><span class="dv">1</span>,<span class="dt">y1=</span><span class="dv">0</span>)
graphics<span class="op">::</span><span class="kw">segments</span>(<span class="dt">x0=</span><span class="dv">0</span>,<span class="dt">x1=</span>n<span class="op">+</span><span class="dv">1</span>,<span class="dt">y0=</span>n<span class="op">+</span><span class="dv">1</span>,<span class="dt">y1=</span><span class="dv">0</span>,<span class="dt">lty=</span><span class="dv">2</span>)

graphics<span class="op">::</span><span class="kw">mtext</span>(<span class="dt">text=</span>cancer,<span class="dt">side=</span><span class="dv">3</span>,<span class="dt">at=</span><span class="dv">1</span><span class="op">:</span>n,<span class="dt">las=</span><span class="dv">2</span>,<span class="dt">cex=</span><span class="fl">0.7</span>)
graphics<span class="op">::</span><span class="kw">mtext</span>(<span class="dt">text=</span>cancer,<span class="dt">side=</span><span class="dv">4</span>,<span class="dt">at=</span>n<span class="op">:</span><span class="dv">1</span>,<span class="dt">las=</span><span class="dv">2</span>,<span class="dt">cex=</span><span class="fl">0.7</span>)

}</code></pre></div>
<hr />
<hr />
<hr />
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
